{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "demanding-space",
   "metadata": {},
   "source": [
    "# Introduction to CPTAC\n",
    "This tutorial will introduce you to the CPTAC dataset through a Python package called `cptac` that was written to access the CPTAC data. This package is similar to `TCGAbiolinks`. The information in `cptac` is stored using `pandas` dataframes. For those curious, the CPTAC documentation is contained [here](https://pypi.org/project/cptac/)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "thick-choice",
   "metadata": {},
   "source": [
    "### Installing and Importing `cptac`\n",
    "Just as in R, we need to install before we can load in `cptac` to our current python environment. If you have not done so already, we need to have installed the package (you can check if it's installed by running the `import cptac` code block below). If you get an error, you will have to take the following steps:\n",
    "\n",
    "1. Shut down your jupyter notebook kernel in the terminal with CTRL+C.\n",
    "2. Make sure that you have deactivated your base environment and activated your `qbio_student` environment.\n",
    "3. Run `pip install cptac`.\n",
    "4. Open the tutorial again through jupyter."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "viral-shelf",
   "metadata": {},
   "source": [
    "### Start exploring CPTAC with `cptac`\n",
    "Similar to `TCGAbiolinks`, we need to load the package and download the data:\n",
    "\n",
    "1. Import the package (`cptac`).\n",
    "2. Look at the data sets available to us with the `list_datasets()` function. As a reminder, remember the package_name.function() syntax!\n",
    "3. Download the one of interest (named `Colon`).\n",
    "4. Load it into our python environment.\n",
    "\n",
    "You might get an error message warning that a method is deprecated -- you can ignore this!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2d34ed3a",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# 1. Import cptac\n",
    "import cptac\n",
    "\n",
    "# 2. Examine the data sets available with list_datasets()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "exceptional-integration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Formatting dataframes...                  \r"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/Users/peterchong/opt/miniconda3/envs/qbio_student/lib/python3.9/site-packages/cptac/colon.py:140: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  prot_combined = prot_tumor.append(prot_normal)\n",
      "/Users/peterchong/opt/miniconda3/envs/qbio_student/lib/python3.9/site-packages/cptac/colon.py:153: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.\n",
      "  phos_combined = phos_tumor.append(phos_normal)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                        \r"
     ]
    }
   ],
   "source": [
    "# 3. Download the colorectal data set -- fill in dataset!\n",
    "cptac.download(dataset=\"Colon\")\n",
    "\n",
    "# 4. Load the colon cancer data set\n",
    "colon = cptac.Colon()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "running-growth",
   "metadata": {},
   "source": [
    "Now, we have our `colon` object containing a bunch of data. We can use the `list_data()` function the data available for the breast cancer dataset. \n",
    "\n",
    "Notice that in addition to the proteomics data, we can also accesss the accompanying transcriptomics, CNV, etc. Additionally, we will focus on the broad proteomics data; however, acetylproteomics and phosphoproteomics are also interesting aspects to explore. These data examine \"post-translational modifications\", in this case the addition of acetyl or phosphate groups to proteins after they are translated, respectively. This data can give insight into which proteins are being used in which cellular pathways.\n",
    "\n",
    "Finally, one last caveat with the data is that many proteins can't be identified very well, so they won't appear as a column (for example, APC isn't in the proteomics data set)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "857a760c",
   "metadata": {},
   "source": [
    "**Exercise**. Call the `list_data()` function from the `colon` object to examine the different data contained in the `colon` object. Much like calling functions from a package, we can use the `colon.function()` syntax."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "71758797",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Below are the dataframes contained in this dataset and their dimensions:\n",
      "\n",
      "clinical\n",
      "\t210 rows\n",
      "\t17 columns\n",
      "CNV\n",
      "\t105 rows\n",
      "\t25187 columns\n",
      "derived_molecular\n",
      "\t110 rows\n",
      "\t11 columns\n",
      "followup\n",
      "\t205 rows\n",
      "\t71 columns\n",
      "miRNA\n",
      "\t105 rows\n",
      "\t989 columns\n",
      "phosphoproteomics\n",
      "\t197 rows\n",
      "\t31339 columns\n",
      "proteomics\n",
      "\t197 rows\n",
      "\t8067 columns\n",
      "somatic_mutation\n",
      "\t93717 rows\n",
      "\t3 columns\n",
      "somatic_mutation_binary\n",
      "\t106 rows\n",
      "\t86026 columns\n",
      "transcriptomics\n",
      "\t106 rows\n",
      "\t13482 columns\n"
     ]
    }
   ],
   "source": [
    "# Run the list_data() function here\n",
    "colon.list_data()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "323f288d",
   "metadata": {},
   "source": [
    "**Exercise.** Now, we want to access the proteomics data. Use the `get_proteomics()` function to store the proteomics data into a separate data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "abf63378",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1BG</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>AAMDC</th>\n",
       "      <th>AAMP</th>\n",
       "      <th>AAR2</th>\n",
       "      <th>...</th>\n",
       "      <th>ZNHIT6</th>\n",
       "      <th>ZNRD1</th>\n",
       "      <th>ZNRF2</th>\n",
       "      <th>ZPR1</th>\n",
       "      <th>ZRANB2</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>-1.1000</td>\n",
       "      <td>0.318</td>\n",
       "      <td>-0.487</td>\n",
       "      <td>0.0995</td>\n",
       "      <td>0.1550</td>\n",
       "      <td>0.1690</td>\n",
       "      <td>0.0653</td>\n",
       "      <td>-0.147</td>\n",
       "      <td>0.1140</td>\n",
       "      <td>0.3410</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0384</td>\n",
       "      <td>0.0221</td>\n",
       "      <td>0.2500</td>\n",
       "      <td>0.0869</td>\n",
       "      <td>0.0331</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.7330</td>\n",
       "      <td>-0.2650</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>-1.1200</td>\n",
       "      <td>-0.441</td>\n",
       "      <td>-0.347</td>\n",
       "      <td>-0.0029</td>\n",
       "      <td>0.0957</td>\n",
       "      <td>0.3960</td>\n",
       "      <td>-0.0363</td>\n",
       "      <td>-0.549</td>\n",
       "      <td>0.2200</td>\n",
       "      <td>0.2480</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.324</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.3940</td>\n",
       "      <td>0.0846</td>\n",
       "      <td>0.0010</td>\n",
       "      <td>-0.3450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.6580</td>\n",
       "      <td>0.0052</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>-1.2000</td>\n",
       "      <td>0.160</td>\n",
       "      <td>-1.850</td>\n",
       "      <td>0.1190</td>\n",
       "      <td>-0.0924</td>\n",
       "      <td>0.0187</td>\n",
       "      <td>-0.2140</td>\n",
       "      <td>0.328</td>\n",
       "      <td>-0.2820</td>\n",
       "      <td>-0.3480</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.3840</td>\n",
       "      <td>-0.1680</td>\n",
       "      <td>0.3570</td>\n",
       "      <td>-0.3250</td>\n",
       "      <td>0.3490</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8210</td>\n",
       "      <td>0.2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>-1.8900</td>\n",
       "      <td>0.112</td>\n",
       "      <td>-0.329</td>\n",
       "      <td>0.6700</td>\n",
       "      <td>0.1160</td>\n",
       "      <td>0.3130</td>\n",
       "      <td>-0.2380</td>\n",
       "      <td>-0.274</td>\n",
       "      <td>-0.5540</td>\n",
       "      <td>0.2700</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0656</td>\n",
       "      <td>-0.2950</td>\n",
       "      <td>0.4630</td>\n",
       "      <td>-0.4480</td>\n",
       "      <td>0.0493</td>\n",
       "      <td>-0.9040</td>\n",
       "      <td>-0.1580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO014</th>\n",
       "      <td>-0.5230</td>\n",
       "      <td>-0.248</td>\n",
       "      <td>-0.638</td>\n",
       "      <td>0.2890</td>\n",
       "      <td>0.3780</td>\n",
       "      <td>0.8220</td>\n",
       "      <td>0.0018</td>\n",
       "      <td>-1.080</td>\n",
       "      <td>0.4980</td>\n",
       "      <td>-0.1200</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0322</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0467</td>\n",
       "      <td>0.2170</td>\n",
       "      <td>0.2360</td>\n",
       "      <td>0.2870</td>\n",
       "      <td>0.3460</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.8120</td>\n",
       "      <td>-0.0273</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21CO006.N</th>\n",
       "      <td>-0.5950</td>\n",
       "      <td>0.430</td>\n",
       "      <td>0.727</td>\n",
       "      <td>0.0141</td>\n",
       "      <td>0.2820</td>\n",
       "      <td>0.0105</td>\n",
       "      <td>0.2780</td>\n",
       "      <td>-0.293</td>\n",
       "      <td>0.0320</td>\n",
       "      <td>0.0128</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.1870</td>\n",
       "      <td>-0.1460</td>\n",
       "      <td>-0.0206</td>\n",
       "      <td>-0.2410</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.4490</td>\n",
       "      <td>0.3200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21CO007.N</th>\n",
       "      <td>0.9770</td>\n",
       "      <td>-0.379</td>\n",
       "      <td>0.661</td>\n",
       "      <td>-0.2120</td>\n",
       "      <td>0.1320</td>\n",
       "      <td>-0.5090</td>\n",
       "      <td>0.4290</td>\n",
       "      <td>1.210</td>\n",
       "      <td>-0.2360</td>\n",
       "      <td>0.4620</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2600</td>\n",
       "      <td>-0.2020</td>\n",
       "      <td>0.2860</td>\n",
       "      <td>-0.1820</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.6200</td>\n",
       "      <td>0.0584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO004.N</th>\n",
       "      <td>0.0542</td>\n",
       "      <td>0.406</td>\n",
       "      <td>1.070</td>\n",
       "      <td>-0.0312</td>\n",
       "      <td>0.0870</td>\n",
       "      <td>0.0428</td>\n",
       "      <td>-0.0950</td>\n",
       "      <td>0.447</td>\n",
       "      <td>0.0081</td>\n",
       "      <td>-0.0947</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.3180</td>\n",
       "      <td>-0.3760</td>\n",
       "      <td>-0.1660</td>\n",
       "      <td>-0.1310</td>\n",
       "      <td>-0.7450</td>\n",
       "      <td>NaN</td>\n",
       "      <td>-0.0342</td>\n",
       "      <td>0.3270</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO006.N</th>\n",
       "      <td>0.5840</td>\n",
       "      <td>-0.691</td>\n",
       "      <td>-0.266</td>\n",
       "      <td>-0.1850</td>\n",
       "      <td>-0.8500</td>\n",
       "      <td>-0.2930</td>\n",
       "      <td>0.5470</td>\n",
       "      <td>0.174</td>\n",
       "      <td>0.0202</td>\n",
       "      <td>0.3160</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0554</td>\n",
       "      <td>0.1810</td>\n",
       "      <td>-0.0940</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.2000</td>\n",
       "      <td>0.1550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27CO004.N</th>\n",
       "      <td>0.9620</td>\n",
       "      <td>-0.171</td>\n",
       "      <td>0.533</td>\n",
       "      <td>0.1370</td>\n",
       "      <td>-0.1840</td>\n",
       "      <td>-0.1670</td>\n",
       "      <td>0.1610</td>\n",
       "      <td>0.636</td>\n",
       "      <td>-0.1800</td>\n",
       "      <td>-0.0453</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1210</td>\n",
       "      <td>0.2070</td>\n",
       "      <td>-0.2680</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.5170</td>\n",
       "      <td>0.1370</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>197 rows × 8067 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name          A1BG   A1CF    A2M    AAAS    AACS   AAGAB    AAK1  AAMDC  \\\n",
       "Patient_ID                                                                \n",
       "01CO005    -1.1000  0.318 -0.487  0.0995  0.1550  0.1690  0.0653 -0.147   \n",
       "01CO006    -1.1200 -0.441 -0.347 -0.0029  0.0957  0.3960 -0.0363 -0.549   \n",
       "01CO008    -1.2000  0.160 -1.850  0.1190 -0.0924  0.0187 -0.2140  0.328   \n",
       "01CO013    -1.8900  0.112 -0.329  0.6700  0.1160  0.3130 -0.2380 -0.274   \n",
       "01CO014    -0.5230 -0.248 -0.638  0.2890  0.3780  0.8220  0.0018 -1.080   \n",
       "...            ...    ...    ...     ...     ...     ...     ...    ...   \n",
       "21CO006.N  -0.5950  0.430  0.727  0.0141  0.2820  0.0105  0.2780 -0.293   \n",
       "21CO007.N   0.9770 -0.379  0.661 -0.2120  0.1320 -0.5090  0.4290  1.210   \n",
       "22CO004.N   0.0542  0.406  1.070 -0.0312  0.0870  0.0428 -0.0950  0.447   \n",
       "22CO006.N   0.5840 -0.691 -0.266 -0.1850 -0.8500 -0.2930  0.5470  0.174   \n",
       "27CO004.N   0.9620 -0.171  0.533  0.1370 -0.1840 -0.1670  0.1610  0.636   \n",
       "\n",
       "Name          AAMP    AAR2  ...  ZNHIT6  ZNRD1   ZNRF2    ZPR1  ZRANB2  \\\n",
       "Patient_ID                  ...                                          \n",
       "01CO005     0.1140  0.3410  ...     NaN    NaN  0.0384  0.0221  0.2500   \n",
       "01CO006     0.2200  0.2480  ...     NaN  0.324     NaN -0.3940  0.0846   \n",
       "01CO008    -0.2820 -0.3480  ...     NaN    NaN -0.3840 -0.1680  0.3570   \n",
       "01CO013    -0.5540  0.2700  ...     NaN    NaN     NaN -0.0656 -0.2950   \n",
       "01CO014     0.4980 -0.1200  ...  0.0322    NaN  0.0467  0.2170  0.2360   \n",
       "...            ...     ...  ...     ...    ...     ...     ...     ...   \n",
       "21CO006.N   0.0320  0.0128  ...     NaN    NaN -0.1870 -0.1460 -0.0206   \n",
       "21CO007.N  -0.2360  0.4620  ...     NaN    NaN     NaN  0.2600 -0.2020   \n",
       "22CO004.N   0.0081 -0.0947  ...     NaN    NaN -0.3180 -0.3760 -0.1660   \n",
       "22CO006.N   0.0202  0.3160  ...     NaN    NaN     NaN  0.0554  0.1810   \n",
       "27CO004.N  -0.1800 -0.0453  ...     NaN    NaN     NaN  0.1210  0.2070   \n",
       "\n",
       "Name          ZW10  ZWILCH   ZWINT     ZYX   ZZEF1  \n",
       "Patient_ID                                          \n",
       "01CO005     0.0869  0.0331     NaN -0.7330 -0.2650  \n",
       "01CO006     0.0010 -0.3450     NaN -0.6580  0.0052  \n",
       "01CO008    -0.3250  0.3490     NaN -0.8210  0.2000  \n",
       "01CO013     0.4630 -0.4480  0.0493 -0.9040 -0.1580  \n",
       "01CO014     0.2870  0.3460     NaN -0.8120 -0.0273  \n",
       "...            ...     ...     ...     ...     ...  \n",
       "21CO006.N  -0.2410     NaN     NaN -0.4490  0.3200  \n",
       "21CO007.N   0.2860 -0.1820     NaN  1.6200  0.0584  \n",
       "22CO004.N  -0.1310 -0.7450     NaN -0.0342  0.3270  \n",
       "22CO006.N  -0.0940     NaN     NaN  1.2000  0.1550  \n",
       "27CO004.N  -0.2680     NaN     NaN  0.5170  0.1370  \n",
       "\n",
       "[197 rows x 8067 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "protein_data = colon.get_proteomics() # get the proteomics data\n",
    "protein_data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "changed-leather",
   "metadata": {},
   "source": [
    "## Analysis\n",
    "We'll be analyzing the proteomics data with `numpy` and `pandas`, so we'll have to load them in as usual.\n",
    "\n",
    "**Exercise**. Import `numpy` and `pandas`. Be sure to assign them their usual shortened names!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "adapted-union",
   "metadata": {},
   "outputs": [],
   "source": [
    "# import packages here\n",
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "promotional-ivory",
   "metadata": {},
   "source": [
    "### Short Pandas Review\n",
    "We can access the rows/column names of a data frame in two ways:\n",
    "\n",
    "1. We can get the row and column names together in a list with the `axes` attribute.\n",
    "2. We can get the row or column names individually in the `index` (row) or `columns` (column) attribute.\n",
    "\n",
    "\n",
    "**Exercise.** Print the `axes`, `index`, and `columns` of the `protein_data` object. Then, determine the dimensions of the `protein_data`, stored in the `shape` attribute. Finally, answer the following questions:\n",
    "\n",
    "- How many rows and columns are there in the data frame?\n",
    "- What do the rows and columns represent?\n",
    "- How many patients are there?\n",
    "- What about proteins?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "expired-reach",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Index(['01CO005', '01CO006', '01CO008', '01CO013', '01CO014', '01CO015',\n",
      "       '01CO019', '01CO022', '05CO002', '05CO003',\n",
      "       ...\n",
      "       '20CO001.N', '20CO003.N', '20CO004.N', '20CO006.N', '20CO007.N',\n",
      "       '21CO006.N', '21CO007.N', '22CO004.N', '22CO006.N', '27CO004.N'],\n",
      "      dtype='object', name='Patient_ID', length=197), Index(['A1BG', 'A1CF', 'A2M', 'AAAS', 'AACS', 'AAGAB', 'AAK1', 'AAMDC', 'AAMP',\n",
      "       'AAR2',\n",
      "       ...\n",
      "       'ZNHIT6', 'ZNRD1', 'ZNRF2', 'ZPR1', 'ZRANB2', 'ZW10', 'ZWILCH', 'ZWINT',\n",
      "       'ZYX', 'ZZEF1'],\n",
      "      dtype='object', name='Name', length=8067)]\n"
     ]
    }
   ],
   "source": [
    "## Explore the data frame here\n",
    "print(protein_data.axes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "add2d694",
   "metadata": {},
   "source": [
    "As for indexing, recall that `loc[]` accesses values by name, and `iloc[]` accesses values by index. `iloc[]` won't be too necessary for this data.\n",
    "\n",
    "**Exercise.** Access the following:\n",
    "1. Print the relative fold change of patient 01CO014 in the `CTNNB1` (beta-catenin) protein. Then, interpret this value -- does this patient have higher or lower protein levels than the reference? How much different is it? (Hint: use `loc[]`. Be sure to specify both the rows and the columns!)\n",
    "2. Access all values in the `COL4A1` column. Because this is a column, there's many ways to do so -- can you think of at least two?\n",
    "3. Finally, use boolean indexing to access the patient IDs with high (≥1) `KRAS` expression:\n",
    "    a. Think about if KRAS expression information will be contained in a row or a column!\n",
    "    b. Create a boolean mask with the row/column that contains the expression information of KRAS.\n",
    "    c. Apply the mask to the row/column names that contains the patient IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "eda8c90d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['05CO028.N'], dtype='object', name='Patient_ID')\n"
     ]
    }
   ],
   "source": [
    "# 1. Access cell of patient 01CO014, for the gene CTNNB1\n",
    "part1 = protein_data.loc[\"01CO014\", \"CTNNB1\"]\n",
    "assert(part1 == -0.0431)\n",
    "\n",
    "# 2. Access all the values in the COL4A1 column\n",
    "part2 = protein_data.loc[:,\"COL4A1\"]\n",
    "assert(len(part2) == 197)\n",
    "\n",
    "# 3. Get the patients with high KRAS expression\n",
    "KRAS_mask = protein_data[\"KRAS\"] > 1\n",
    "part3 = protein_data.index[KRAS_mask] #index is used to get rows in pandas\n",
    "print(part3)\n",
    "assert(part3 == \"05CO028.N\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "forbidden-advertiser",
   "metadata": {},
   "source": [
    "### Additional data in cptac\n",
    "You can also access other data using the `cptac` python package for the same patients, using the corresponding `get` function.\n",
    "\n",
    "**Exercise**. Access the RNA (`transcriptomics`) and `clinical` data from the colon object. The function name is very similar in syntax to the `get_proteomics()` function from before!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "13e51b2e",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_data = colon.get_transcriptomics()\n",
    "clinical_data = colon.get_clinical()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad3e3bd2",
   "metadata": {},
   "source": [
    "**Exercise**. Examine these dataframes. In particular:\n",
    "\n",
    "- What do the values in the `rna_data` data frame represent?\n",
    "- Compare the dimensions of the `protein_data`, `rna_data`, and `clinical_data`. Are they equal?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "a54f52a4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADACP1</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAED1</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>10.1984</td>\n",
       "      <td>13.8117</td>\n",
       "      <td>7.0224</td>\n",
       "      <td>9.4115</td>\n",
       "      <td>9.7432</td>\n",
       "      <td>4.5850</td>\n",
       "      <td>8.4635</td>\n",
       "      <td>7.6221</td>\n",
       "      <td>11.2336</td>\n",
       "      <td>8.3219</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1046</td>\n",
       "      <td>10.5765</td>\n",
       "      <td>10.5294</td>\n",
       "      <td>7.0553</td>\n",
       "      <td>8.4263</td>\n",
       "      <td>8.6865</td>\n",
       "      <td>10.0307</td>\n",
       "      <td>9.1699</td>\n",
       "      <td>10.2784</td>\n",
       "      <td>10.4263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>10.1586</td>\n",
       "      <td>12.6434</td>\n",
       "      <td>6.7279</td>\n",
       "      <td>8.9972</td>\n",
       "      <td>9.9084</td>\n",
       "      <td>4.7004</td>\n",
       "      <td>8.2046</td>\n",
       "      <td>7.0768</td>\n",
       "      <td>10.5362</td>\n",
       "      <td>9.6402</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6883</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>9.2808</td>\n",
       "      <td>7.1799</td>\n",
       "      <td>9.2946</td>\n",
       "      <td>9.1799</td>\n",
       "      <td>9.8408</td>\n",
       "      <td>10.5622</td>\n",
       "      <td>10.4512</td>\n",
       "      <td>10.4305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>9.2119</td>\n",
       "      <td>15.0040</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>9.7056</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>8.4676</td>\n",
       "      <td>7.1293</td>\n",
       "      <td>10.9337</td>\n",
       "      <td>9.5661</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8424</td>\n",
       "      <td>9.8611</td>\n",
       "      <td>10.2107</td>\n",
       "      <td>7.6582</td>\n",
       "      <td>8.7415</td>\n",
       "      <td>9.0741</td>\n",
       "      <td>10.3309</td>\n",
       "      <td>10.0512</td>\n",
       "      <td>10.2703</td>\n",
       "      <td>10.5430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>10.7322</td>\n",
       "      <td>13.9284</td>\n",
       "      <td>4.9542</td>\n",
       "      <td>9.1497</td>\n",
       "      <td>8.9425</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.4179</td>\n",
       "      <td>7.7074</td>\n",
       "      <td>10.7194</td>\n",
       "      <td>9.3376</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6812</td>\n",
       "      <td>9.0795</td>\n",
       "      <td>9.5118</td>\n",
       "      <td>8.9159</td>\n",
       "      <td>9.9542</td>\n",
       "      <td>10.0168</td>\n",
       "      <td>10.3365</td>\n",
       "      <td>9.9307</td>\n",
       "      <td>10.2538</td>\n",
       "      <td>10.8925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>9.7160</td>\n",
       "      <td>12.9963</td>\n",
       "      <td>6.3576</td>\n",
       "      <td>9.6202</td>\n",
       "      <td>9.2738</td>\n",
       "      <td>4.1699</td>\n",
       "      <td>7.0980</td>\n",
       "      <td>5.9069</td>\n",
       "      <td>10.2252</td>\n",
       "      <td>9.7944</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1344</td>\n",
       "      <td>9.3174</td>\n",
       "      <td>10.0238</td>\n",
       "      <td>7.4429</td>\n",
       "      <td>8.7381</td>\n",
       "      <td>9.6165</td>\n",
       "      <td>10.3848</td>\n",
       "      <td>11.3066</td>\n",
       "      <td>11.2174</td>\n",
       "      <td>10.7616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21CO007</th>\n",
       "      <td>10.0768</td>\n",
       "      <td>13.4762</td>\n",
       "      <td>6.4594</td>\n",
       "      <td>9.6018</td>\n",
       "      <td>9.1573</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>6.2095</td>\n",
       "      <td>10.9159</td>\n",
       "      <td>10.9315</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1293</td>\n",
       "      <td>9.5622</td>\n",
       "      <td>10.2958</td>\n",
       "      <td>7.5314</td>\n",
       "      <td>8.3707</td>\n",
       "      <td>8.8202</td>\n",
       "      <td>9.8234</td>\n",
       "      <td>9.4136</td>\n",
       "      <td>9.7245</td>\n",
       "      <td>10.6777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO004</th>\n",
       "      <td>10.9293</td>\n",
       "      <td>12.6131</td>\n",
       "      <td>5.5236</td>\n",
       "      <td>9.3332</td>\n",
       "      <td>9.2877</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.3663</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>10.9031</td>\n",
       "      <td>8.8361</td>\n",
       "      <td>...</td>\n",
       "      <td>9.9687</td>\n",
       "      <td>9.5736</td>\n",
       "      <td>10.5108</td>\n",
       "      <td>8.4346</td>\n",
       "      <td>9.0768</td>\n",
       "      <td>8.8857</td>\n",
       "      <td>10.2715</td>\n",
       "      <td>9.4737</td>\n",
       "      <td>9.6511</td>\n",
       "      <td>10.9114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO006</th>\n",
       "      <td>9.2432</td>\n",
       "      <td>14.2620</td>\n",
       "      <td>7.5546</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>8.7039</td>\n",
       "      <td>5.5546</td>\n",
       "      <td>7.3129</td>\n",
       "      <td>8.0389</td>\n",
       "      <td>11.1592</td>\n",
       "      <td>9.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8903</td>\n",
       "      <td>9.9972</td>\n",
       "      <td>10.1923</td>\n",
       "      <td>7.4998</td>\n",
       "      <td>8.1447</td>\n",
       "      <td>8.7582</td>\n",
       "      <td>10.1497</td>\n",
       "      <td>10.4062</td>\n",
       "      <td>10.2491</td>\n",
       "      <td>10.3902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24CO005</th>\n",
       "      <td>8.6546</td>\n",
       "      <td>13.0250</td>\n",
       "      <td>8.0498</td>\n",
       "      <td>10.2228</td>\n",
       "      <td>9.3729</td>\n",
       "      <td>3.1699</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>8.3399</td>\n",
       "      <td>11.0566</td>\n",
       "      <td>8.0980</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8933</td>\n",
       "      <td>10.2621</td>\n",
       "      <td>10.8727</td>\n",
       "      <td>6.2479</td>\n",
       "      <td>7.8392</td>\n",
       "      <td>8.1849</td>\n",
       "      <td>9.6917</td>\n",
       "      <td>8.3264</td>\n",
       "      <td>7.8704</td>\n",
       "      <td>10.3740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27CO004</th>\n",
       "      <td>9.8025</td>\n",
       "      <td>13.1139</td>\n",
       "      <td>7.9069</td>\n",
       "      <td>9.7582</td>\n",
       "      <td>9.2479</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.7682</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>11.0049</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6439</td>\n",
       "      <td>9.8857</td>\n",
       "      <td>10.3049</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>8.6110</td>\n",
       "      <td>9.1624</td>\n",
       "      <td>9.3106</td>\n",
       "      <td>11.0532</td>\n",
       "      <td>9.6777</td>\n",
       "      <td>9.6900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 13482 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "Name           A1CF      A2M  A4GALT     AAAS    AACS  AADACP1   AADAT  \\\n",
       "Patient_ID                                                               \n",
       "01CO001     10.1984  13.8117  7.0224   9.4115  9.7432   4.5850  8.4635   \n",
       "01CO005     10.1586  12.6434  6.7279   8.9972  9.9084   4.7004  8.2046   \n",
       "01CO006      9.2119  15.0040  6.8580   9.7056  9.1997   5.0000  8.4676   \n",
       "01CO008     10.7322  13.9284  4.9542   9.1497  8.9425   6.0224  7.4179   \n",
       "01CO013      9.7160  12.9963  6.3576   9.6202  9.2738   4.1699  7.0980   \n",
       "...             ...      ...     ...      ...     ...      ...     ...   \n",
       "21CO007     10.0768  13.4762  6.4594   9.6018  9.1573   5.0000  7.6865   \n",
       "22CO004     10.9293  12.6131  5.5236   9.3332  9.2877   5.0000  7.3663   \n",
       "22CO006      9.2432  14.2620  7.5546   9.3264  8.7039   5.5546  7.3129   \n",
       "24CO005      8.6546  13.0250  8.0498  10.2228  9.3729   3.1699  7.0444   \n",
       "27CO004      9.8025  13.1139  7.9069   9.7582  9.2479   6.0224  7.7682   \n",
       "\n",
       "Name         AAED1    AAGAB     AAK1  ...     ZW10   ZWILCH    ZWINT    ZXDA  \\\n",
       "Patient_ID                            ...                                      \n",
       "01CO001     7.6221  11.2336   8.3219  ...  10.1046  10.5765  10.5294  7.0553   \n",
       "01CO005     7.0768  10.5362   9.6402  ...   9.6883   9.3264   9.2808  7.1799   \n",
       "01CO006     7.1293  10.9337   9.5661  ...   9.8424   9.8611  10.2107  7.6582   \n",
       "01CO008     7.7074  10.7194   9.3376  ...   9.6812   9.0795   9.5118  8.9159   \n",
       "01CO013     5.9069  10.2252   9.7944  ...  10.1344   9.3174  10.0238  7.4429   \n",
       "...            ...      ...      ...  ...      ...      ...      ...     ...   \n",
       "21CO007     6.2095  10.9159  10.9315  ...  10.1293   9.5622  10.2958  7.5314   \n",
       "22CO004     6.8580  10.9031   8.8361  ...   9.9687   9.5736  10.5108  8.4346   \n",
       "22CO006     8.0389  11.1592   9.2597  ...   9.8903   9.9972  10.1923  7.4998   \n",
       "24CO005     8.3399  11.0566   8.0980  ...   9.8933  10.2621  10.8727  6.2479   \n",
       "27CO004     7.0444  11.0049   9.1997  ...   9.6439   9.8857  10.3049  7.6865   \n",
       "\n",
       "Name          ZXDB     ZXDC   ZYG11B      ZYX    ZZEF1     ZZZ3  \n",
       "Patient_ID                                                       \n",
       "01CO001     8.4263   8.6865  10.0307   9.1699  10.2784  10.4263  \n",
       "01CO005     9.2946   9.1799   9.8408  10.5622  10.4512  10.4305  \n",
       "01CO006     8.7415   9.0741  10.3309  10.0512  10.2703  10.5430  \n",
       "01CO008     9.9542  10.0168  10.3365   9.9307  10.2538  10.8925  \n",
       "01CO013     8.7381   9.6165  10.3848  11.3066  11.2174  10.7616  \n",
       "...            ...      ...      ...      ...      ...      ...  \n",
       "21CO007     8.3707   8.8202   9.8234   9.4136   9.7245  10.6777  \n",
       "22CO004     9.0768   8.8857  10.2715   9.4737   9.6511  10.9114  \n",
       "22CO006     8.1447   8.7582  10.1497  10.4062  10.2491  10.3902  \n",
       "24CO005     7.8392   8.1849   9.6917   8.3264   7.8704  10.3740  \n",
       "27CO004     8.6110   9.1624   9.3106  11.0532   9.6777   9.6900  \n",
       "\n",
       "[106 rows x 13482 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_data\n",
    "print(protein_data.shape, rna_data.shape, clinical_data.shape)\n",
    "# explore the data frames!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "4be82967",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([2])"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.intersect1d([1,2], [2,3,4])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6cac2236",
   "metadata": {},
   "source": [
    "We can use the `intersect1d()` function from `numpy` to see which patient barcodes are shared between the data frames. **The reason we might be interested in these patients is that we can compare the transcriptomics to proteomics data.**\n",
    "\n",
    "1. Use `intersect1d()` to create a list that contains the patient IDs for each pair of data frames as mentioned below.\n",
    "2. Use a for loop to print the lengths of these three data frames, and compare them to the number of rows. The for loop skeleton is below. What do you notice about the three lengths? Do the numbers correspond to anything in particular?\n",
    "3. Finally, identify which intersection contains the patient IDs where there is data for all three levels of data (proteomics, transcriptomics, and clinical). How can we tell?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "chemical-location",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The length of 0 is 96\n",
      "The length of 1 is 197\n",
      "The length of 2 is 106\n"
     ]
    }
   ],
   "source": [
    "# 1.\n",
    "name_intersects = [\n",
    "    np.intersect1d(protein_data.index , rna_data.index), # 0. fill in intersecting for protein/rna here\n",
    "    np.intersect1d(protein_data.index , clinical_data.index), # 1. fill in intersecting for protein/clinical here\n",
    "    np.intersect1d(rna_data.index , clinical_data.index), # 2. fill in the intersecting for rna/clinical here\n",
    "]\n",
    "\n",
    "# 2. Print the lengths here\n",
    "for i, intersects in enumerate(name_intersects):\n",
    "    print(f\"The length of {i} is {len(intersects)}\")  # fill in here\n",
    "    \n",
    "# 3. Which comparison contains the patient names that have all three levels?"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46124874",
   "metadata": {},
   "source": [
    "### Examining the clinical data\n",
    "Let's explore the clinical data in more detail.\n",
    "\n",
    "**Exercise**. First, let's use `head()` to glance as to what data is available."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "5b38bbe6",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>Age</th>\n",
       "      <th>CEA</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Lymphatic_Invasion</th>\n",
       "      <th>Mucinous</th>\n",
       "      <th>Perineural_Invasion</th>\n",
       "      <th>Polyps_History</th>\n",
       "      <th>Polyps_Present</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Subsite</th>\n",
       "      <th>Synchronous_Tumors</th>\n",
       "      <th>Tumor.Status</th>\n",
       "      <th>Vascular_Invasion</th>\n",
       "      <th>Vital.Status</th>\n",
       "      <th>pathalogy_N_stage</th>\n",
       "      <th>pathalogy_T_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>729.0</td>\n",
       "      <td>4.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>838.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>904.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>Yes</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>652.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>695.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Sample_Tumor_Normal    Age  CEA  Gender Lymphatic_Invasion  \\\n",
       "Patient_ID                                                              \n",
       "01CO001                  Tumor  729.0  4.9    Male                Yes   \n",
       "01CO005                  Tumor  838.0  1.0  Female                 No   \n",
       "01CO006                  Tumor  904.0  NaN  Female                Yes   \n",
       "01CO008                  Tumor  652.0  NaN  Female                 No   \n",
       "01CO013                  Tumor  695.0  NaN    Male                 No   \n",
       "\n",
       "Name            Mucinous Perineural_Invasion Polyps_History Polyps_Present  \\\n",
       "Patient_ID                                                                   \n",
       "01CO001         Mucinous                  No             No             No   \n",
       "01CO005     Not Mucinous                  No            Yes            Yes   \n",
       "01CO006         Mucinous                 Yes             No             No   \n",
       "01CO008         Mucinous                  No             No            Yes   \n",
       "01CO013     Not Mucinous                  No            Yes             No   \n",
       "\n",
       "Name            Stage           Subsite Synchronous_Tumors Tumor.Status  \\\n",
       "Patient_ID                                                                \n",
       "01CO001     Stage III     Sigmoid Colon                 No   Tumor free   \n",
       "01CO005      Stage II     Sigmoid Colon                 No   Tumor free   \n",
       "01CO006     Stage III   Ascending Colon                Yes   With tumor   \n",
       "01CO008      Stage II  Descending Colon                 No   With tumor   \n",
       "01CO013       Stage I     Sigmoid Colon                 No   Tumor free   \n",
       "\n",
       "Name       Vascular_Invasion Vital.Status pathalogy_N_stage pathalogy_T_stage  \n",
       "Patient_ID                                                                     \n",
       "01CO001                  Yes       Living               N2b               T4a  \n",
       "01CO005                   No     Deceased                N0                T3  \n",
       "01CO006                  Yes       Living               N2b               T4a  \n",
       "01CO008                   No       Living                N0                T3  \n",
       "01CO013                   No       Living                N0                T2  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# call head on the clinical data!\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9e5486b7",
   "metadata": {},
   "source": [
    "**Exercise**. You might notice that the Age column values don't look like years -- they're in months. Divide the values by 12 in the `clinical_data` data frame to turn them into years. Then, use `head()` again to make sure the ages were converted."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "61978ca9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>Name</th>\n",
       "      <th>Sample_Tumor_Normal</th>\n",
       "      <th>Age</th>\n",
       "      <th>CEA</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Lymphatic_Invasion</th>\n",
       "      <th>Mucinous</th>\n",
       "      <th>Perineural_Invasion</th>\n",
       "      <th>Polyps_History</th>\n",
       "      <th>Polyps_Present</th>\n",
       "      <th>Stage</th>\n",
       "      <th>Subsite</th>\n",
       "      <th>Synchronous_Tumors</th>\n",
       "      <th>Tumor.Status</th>\n",
       "      <th>Vascular_Invasion</th>\n",
       "      <th>Vital.Status</th>\n",
       "      <th>pathalogy_N_stage</th>\n",
       "      <th>pathalogy_T_stage</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>60.750000</td>\n",
       "      <td>4.9</td>\n",
       "      <td>Male</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>69.833333</td>\n",
       "      <td>1.0</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Deceased</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>75.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>Yes</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Living</td>\n",
       "      <td>N2b</td>\n",
       "      <td>T4a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>54.333333</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Female</td>\n",
       "      <td>No</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>With tumor</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>Tumor</td>\n",
       "      <td>57.916667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Male</td>\n",
       "      <td>No</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>No</td>\n",
       "      <td>Yes</td>\n",
       "      <td>No</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>No</td>\n",
       "      <td>Tumor free</td>\n",
       "      <td>No</td>\n",
       "      <td>Living</td>\n",
       "      <td>N0</td>\n",
       "      <td>T2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Name       Sample_Tumor_Normal        Age  CEA  Gender Lymphatic_Invasion  \\\n",
       "Patient_ID                                                                  \n",
       "01CO001                  Tumor  60.750000  4.9    Male                Yes   \n",
       "01CO005                  Tumor  69.833333  1.0  Female                 No   \n",
       "01CO006                  Tumor  75.333333  NaN  Female                Yes   \n",
       "01CO008                  Tumor  54.333333  NaN  Female                 No   \n",
       "01CO013                  Tumor  57.916667  NaN    Male                 No   \n",
       "\n",
       "Name            Mucinous Perineural_Invasion Polyps_History Polyps_Present  \\\n",
       "Patient_ID                                                                   \n",
       "01CO001         Mucinous                  No             No             No   \n",
       "01CO005     Not Mucinous                  No            Yes            Yes   \n",
       "01CO006         Mucinous                 Yes             No             No   \n",
       "01CO008         Mucinous                  No             No            Yes   \n",
       "01CO013     Not Mucinous                  No            Yes             No   \n",
       "\n",
       "Name            Stage           Subsite Synchronous_Tumors Tumor.Status  \\\n",
       "Patient_ID                                                                \n",
       "01CO001     Stage III     Sigmoid Colon                 No   Tumor free   \n",
       "01CO005      Stage II     Sigmoid Colon                 No   Tumor free   \n",
       "01CO006     Stage III   Ascending Colon                Yes   With tumor   \n",
       "01CO008      Stage II  Descending Colon                 No   With tumor   \n",
       "01CO013       Stage I     Sigmoid Colon                 No   Tumor free   \n",
       "\n",
       "Name       Vascular_Invasion Vital.Status pathalogy_N_stage pathalogy_T_stage  \n",
       "Patient_ID                                                                     \n",
       "01CO001                  Yes       Living               N2b               T4a  \n",
       "01CO005                   No     Deceased                N0                T3  \n",
       "01CO006                  Yes       Living               N2b               T4a  \n",
       "01CO008                   No       Living                N0                T3  \n",
       "01CO013                   No       Living                N0                T2  "
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinical_data[\"Age\"] /= 12\n",
    "clinical_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d704ebf5",
   "metadata": {},
   "source": [
    "There are actually control (non-cancer) tissue samples in the data! Let's show this using the `unique()` function from numpy.\n",
    "\n",
    "**Exercise.** Call the `unique()` function on the `Sample_Tumor_Normal` column. The function will return the unique values in the column in a sorted order -- this is super useful for examining categorical variables, like tumor stage and tumor status, for example."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "416c93c1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Normal', 'Tumor'], dtype=object)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get the levels of the Sample_Tumor_Normal column with unique()\n",
    "np.unique(clinical_data[\"Sample_Tumor_Normal\"])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5ff472",
   "metadata": {},
   "source": [
    "**Exercise.** Let's examine how tumor stage varies as a function of age:\n",
    "\n",
    "1. Import the plotting libraries `matplotlib.pyplot` and `seaborn` (use the standard abbreviations presented last time).\n",
    "2. Create an `age_category` column in `clinical_data`. As before, we'll define \"Young\" as under 50, and \"Old\" as 50 and older. (Hint: use the `where()` function from numpy, the equivalent to `ifelse()`).\n",
    "3.  The \"Stage\" information for these patients is NaN (the pandas version of NA). Use the `isna()` function from pandas and boolean indexing (i.e. `where()` from numpy again) to replace these NaN values in the `Stage` column with \"No cancer\".\n",
    "3. Using nested for loop and boolean indexing, count the number of old and young patients that have No cancer, Stage I, Stage II, Stage III, and Stage IV cancer. \n",
    "4. Compare your results to the barplot. Do your numbers make sense?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "6e84c387",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Old\n",
      "There are 100 patients with no cancer, 12 patients with Stage I cancer, 40 patients with Stage II cancer, 43 patients with Stage III cancer, and 8 patients with IV cancer.\n",
      "Young\n",
      "There are 0 patients with no cancer, 0 patients with Stage I cancer, 2 patients with Stage II cancer, 5 patients with Stage III cancer, and 0 patients with IV cancer.\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYUAAAEHCAYAAABBW1qbAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAbOUlEQVR4nO3df3RV9Znv8fcjQSK/bCCpA1INjCgaiIFGLwpLA1SEO1YcHfnRoijeC6NgCsvrrc6tA1XpYtZFq6DOurFWqKMggorlSi9KA04YpBBgUH6Jo4hUBgkiQmhqsc/94+zsHiCBk5Bz9knO57VW1jn7e/aP52BWPn733uc55u6IiIgAnBV1ASIikj4UCiIiElIoiIhISKEgIiIhhYKIiIQUCiIiEsqKuoAzkZub6/n5+VGXISLSrFRWVla5e15drzXrUMjPz2f9+vVRlyEi0qyY2Sf1vabTRyIiElIoiIhISKEgIiKhZn1NQURajj/96U/s2bOHmpqaqEtpMbKzs+nWrRutW7dOeBuFgoikhT179tChQwfy8/Mxs6jLafbcnQMHDrBnzx66d++e8HZJO31kZr80s8/N7P24sU5m9paZ7Qwec+Jee9DMPjSzHWZ2fbLqEpH0VFNTQ+fOnRUITcTM6Ny5c4NnXsm8pjAXGHbC2APACnfvCawIljGzy4DRQEGwzTNm1iqJtYlIGlIgNK3G/HsmLRTc/R3gixOGRwDzgufzgJvixhe4+x/d/WPgQ+DKZNUmIlJrxowZFBQUUFhYSFFREWvXruWJJ57g6NGjUZcWiVRfUzjP3fcCuPteM/t2MH4+8G7cenuCsZOY2QRgAsAFF1xwxgV99/5fnfE+ElX5v29P2bFE5PTWrFnD0qVL2bBhA23atKGqqoqvv/6aUaNGMXbsWNq2bRt1iSmXLrek1jXHqfMr4dy9zN2L3b04L6/OT2mLiCRk79695Obm0qZNGwByc3NZtGgRn332GYMGDWLQoEEA3H333RQXF1NQUMC0adPC7d9880169erFwIEDKS0t5YYbbgCgurqa8ePHc8UVV9C3b1+WLFmS+jfXSKkOhX1m1gUgePw8GN8DfCduvW7AZymuTUQyzNChQ/n000+5+OKLueeee1i1ahWlpaV07dqV8vJyysvLgdgppvXr17N582ZWrVrF5s2bqampYeLEiSxbtoyKigr2798f7nfGjBkMHjyYdevWUV5ezv333091dXVUb7NBUh0KbwDjgufjgCVx46PNrI2ZdQd6Ar9LcW0ikmHat29PZWUlZWVl5OXlMWrUKObOnXvSegsXLqRfv3707duXLVu2sHXrVrZv306PHj3C2z3HjBkTrr98+XJmzpxJUVERJSUl1NTUsHv37lS9rTOStGsKZjYfKAFyzWwPMA2YCSw0s7uA3cCtAO6+xcwWAluBY8Akd/8mWbWJiNRq1aoVJSUllJSU0KdPH+bNm3fc6x9//DGzZs1i3bp15OTkcMcdd1BTU4N7nWe4gdhnBBYvXswll1yS7PKbXDLvPhrj7l3cvbW7d3P359z9gLsPcfeeweMXcevPcPe/dvdL3H1ZsuoSEam1Y8cOdu7cGS5v2rSJCy+8kA4dOnD48GEAvvrqK9q1a8e5557Lvn37WLYs9uepV69efPTRR+zatQuAl19+OdzP9ddfz5w5c8Lg2LhxY4re0ZnTJ5pFJGMdOXKEe++9ly+//JKsrCwuuugiysrKmD9/PsOHD6dLly6Ul5fTt29fCgoK6NGjBwMGDADgnHPO4ZlnnmHYsGHk5uZy5ZV/uYv+oYceYsqUKRQWFuLu5Ofns3Tp0qjeZoPYqaZA6a64uNjP9PsUdEuqSHrYtm0bl156adRlNMiRI0do37497s6kSZPo2bMnU6dOjbqs49T172pmle5eXNf66XJLqohIs/Pss89SVFREQUEBhw4dYuLEiVGXdMZ0+khEpJGmTp2adjODM6WZgoiIhBQKIiISUiiIiEhIoSAiIiGFgohIwMy47777wuVZs2Yxffr06AqKgO4+EpG01NSfIUrkc0Jt2rTh1Vdf5cEHHyQ3N7dJj58q33zzDa1aNf47yjRTEBEJZGVlMWHCBH7+85+f9Nonn3zCkCFDKCwsZMiQIXU2uDty5Ah33nknffr0obCwkMWLFwP1t97Oz89n2rRp9OvXjz59+rB9+/ZT7mf58uVcddVV9OvXj1tvvZUjR46E+3n44YcZOHAgr7zyyhn9GygURETiTJo0iRdffJFDhw4dNz558mRuv/12Nm/ezA9/+ENKS0tP2vaRRx7h3HPP5b333mPz5s0MHjwYqLv1dq3c3Fw2bNjA3XffzaxZs+rdT1VVFY8++ihvv/02GzZsoLi4mMcffzzcT3Z2NhUVFYwePfqM3r9CQUQkTseOHbn99tuZPXv2ceNr1qzhBz/4AQC33XYbFRUVJ2379ttvM2nSpHA5JycHqLv1dq2bb74ZgO9+97thc7269vPuu++ydetWBgwYQFFREfPmzeOTTz4J1xk1atQZvvMYXVMQETnBlClT6NevH3feeWe965id/IWR7n7SeH2tt2vVfutbq1atOHbsWL37cXeuu+465s+fX2c97dq1S+zNnYZmCiIiJ+jUqRMjR47kueeeC8euvvpqFixYAMCLL77IwIEDT9pu6NChPPXUU+HywYMH6229fSp17ad///6sXr2aDz/8EICjR4/ywQcfNPo91kehICJSh/vuu4+qqqpwefbs2Tz//PMUFhbywgsv8OSTT560zU9+8hMOHjxI7969ufzyyykvL+fyyy8PW2+PHz8+bL19KnXtJy8vj7lz5zJmzBgKCwvp379/eGG6Kal1tlpni6SF5tg6uzlQ62wREWk0hYKIiIQUCiIiElIoiIhISKEgIiIhhYKIiIQUCiIigRkzZlBQUEBhYSFFRUWsXbsWgCeeeIKjR48m9di7du2id+/eST1GItTmQkTS0u6H+zTp/i74x/dO+fqaNWtYunQpGzZsoE2bNlRVVfH1118DsVAYO3Ysbdu2bdKa0pFmCiIiwN69e8nNzQ17EeXm5tK1a1dmz57NZ599xqBBgxg0aBBQfyvsN998k169ejFw4EBKS0u54YYbAKiurmb8+PFcccUV9O3blyVLlqT+DSZIoSAiQqzf0KeffsrFF1/MPffcw6pVqwAoLS2la9eulJeXU15eDtTdCrumpoaJEyeybNkyKioq2L9/f7jvGTNmMHjwYNatW0d5eTn3338/1dXVkbzP01EoiIgA7du3p7KykrKyMvLy8hg1ahRz586tc926WmFv376dHj160L17dwDGjBkTrr98+XJmzpxJUVERJSUl1NTU1PklPelA1xRERAKtWrWipKSEkpIS+vTpw7x587jjjjuOW6e+Vtin6iPn7ixevJhLLrkkye/gzGmmICIC7Nixg507d4bLmzZt4sILLwSgQ4cOHD58GKDeVti9evXio48+Cr8o5+WXXw73df311zNnzpwwODZu3JiKt9QomimIiBD7XuR7772XL7/8kqysLC666CLKysoAmDBhAsOHD6dLly6Ul5eHrbB79OgRtsI+55xzeOaZZxg2bBi5ublceeWV4b4feughpkyZQmFhIe5Ofn4+S5cujeR9no5aZ6t1tkhaaAmts48cOUL79u1xdyZNmkTPnj2ZOnVqpDU1i9bZZjbVzLaY2ftmNt/Mss2sk5m9ZWY7g8ecKGoTEWmsZ599lqKiIgoKCjh06BATJ06MuqQGS/npIzM7HygFLnP3P5jZQmA0cBmwwt1nmtkDwAPAj1Ndn4hIY02dOjXymcGZiupCcxZwjpllAW2Bz4ARwLzg9XnATdGUJiKSuVIeCu7+e2AWsBvYCxxy9+XAee6+N1hnL/DtVNcmIpLpUh4KwbWCEUB3oCvQzszGNmD7CWa23szWx39iUEREzlwUp4++B3zs7vvd/U/Aq8DVwD4z6wIQPH5e18buXubuxe5enJeXl7KiRUQyQRShsBvob2ZtzcyAIcA24A1gXLDOOCB9O0aJSIuULq2zV65cGTbTS7WU333k7mvNbBGwATgGbATKgPbAQjO7i1hw3Jrq2kQkfQyYM6BJ97f63tWnfF2ts2MiufvI3ae5ey937+3ut7n7H939gLsPcfeeweMXUdQmIplJrbNj1PtIRAS1zq6lUBARQa2za6khnohIQK2zNVMQEQHUOruWZgoiIqh1di21zlbrbJG0oNbZydEsWmeLiLREap0tIiIhtc4WEZEWRaEgIiIhhYKIiIQUCiIiElIoiIgE0rF19ty5c5k8eXJSjx1Pdx+JSFpadc21Tbq/a99ZdcrX1To7RjMFERHUOruWQkFEBLXOrqVQEBFBrbNr6ZqCiEhArbM1UxARAdQ6u5ZmCiIiqHV2LbXOVutskbSg1tnJodbZIiIRUetsEREJqXW2iIi0KAoFEREJKRRERCSkUBARkZBCQUQkkA6ts6urq+ncuTOHDh067vWbbrqJhQsXJrUG0N1HIpKmnrrv1026v8mPff+Ur6dL6+x27doxdOhQXn/9dcaNGwfAoUOHqKio4KWXXkr68TVTEBEhvVpnjxkzhgULFoTLr732GsOGDUtJKCkURERIr9bZw4YNo7KykgMHDgCwYMGC47quJpNCQUSE9GqdffbZZ3PjjTeyaNEiqqqq2LRpE0OHDm3S91ufSK4pmNm3gF8AvQEHxgM7gJeBfGAXMNLdD0ZRn4hkpnRqnT1mzBgeffRR3J0RI0bQunXrxr6tBolqpvAk8Bt37wVcDmwDHgBWuHtPYEWwLCKSEunWOnvQoEHs3LmTp59+OmWnjiCCmYKZdQSuAe4AcPevga/NbARQEqw2D1gJ/DjV9YlIZkq31tlnnXUWt9xyC6+88grXXHNN8t74CVLeOtvMioAyYCuxWUIl8CPg9+7+rbj1Drp7zqn2pdbZIi2HWmcnR3NonZ0F9AP+2d37AtU04FSRmU0ws/Vmtj7+6r6ISNTUOrtx9gB73H1tsLyIWCjsM7Mu7r7XzLoAn9e1sbuXEZtpUFxc3Hy/IUhEWhy1zm4Ed/9P4FMzq70MP4TYqaQ3gHHB2Djg1J/uEBGRJhdVm4t7gRfN7GzgI+BOYgG10MzuAnYDt0ZUm4hExN0xs6jLaDEac804oVAwsxXuPuR0Y4ly901AXRc5GrU/EWn+srOzOXDgAJ07d1YwNAF358CBA2RnZzdou1OGgpllA22BXDPLAWr/S3UEujamUBGRunTr1o09e/agG0iaTnZ2Nt26dWvQNqebKUwEphALgEr+EgpfAU83sD4RkXq1bt06bBEh0TllKLj7k8CTZnavu89JUU0iIhKRhK4puPscM7uaWF+irLjx1H3yS0REki7RC80vAH8NbAK+CYYdUCiIiLQgid6SWgxc5qnuiSEiIimV6IfX3gf+KpmFiIhI9BKdKeQCW83sd8Afawfd/cakVCUiIpFINBSmJ7MIERFJD4nefbQq2YWIiEj0Er376DCxu40AzgZaA9Xu3jFZhYmISOolOlPoEL9sZjcBV9a9toiINFeNap3t7q8Dg5u2FBERiVqip49ujls8i9jnFvSZBRGRFibRu4++H/f8GLALGNHk1YiISKQSvaZwZ7ILERGR6CV0TcHMupnZa2b2uZntM7PFZtawJt0iIpL2Er3Q/Dyx71DuCpwP/DoYExGRFiTRUMhz9+fd/VjwMxfIS2JdIiISgURDocrMxppZq+BnLHAgmYWJiEjqJRoK44GRwH8Ce4G/A3TxWUSkhUn0ltRHgHHufhDAzDoBs4iFhYiItBCJzhQKawMBwN2/APompyQREYlKoqFwlpnl1C4EM4VEZxkiItJMJPqH/THg38xsEbH2FiOBGUmrSkREIpHoJ5p/ZWbriTXBM+Bmd9+a1MpERCTlEj4FFISAgkBEpAVrVOtsERFpmRQKIiISUiiIiEhIt5XKGVl1zbUpO9a176xK2bFEMpVmCiIiElIoiIhIKLJQCLqtbjSzpcFyJzN7y8x2Bo85p9uHiIg0rShnCj8CtsUtPwCscPeewIpgWUREUiiSUAi+yvNvgF/EDY8A5gXP5wE3pbgsEZGMF9VM4QngfwJ/jhs7z933AgSP346gLhGRjJbyW1LN7Abgc3evNLOSRmw/AZgAcMEFFzRtcS3IgDkDUnKcn+muZpEWJYqZwgDgRjPbBSwABpvZvwD7zKwLQPD4eV0bu3uZuxe7e3Fenr4mWkSkKaU8FNz9QXfv5u75wGjgt+4+FngDGBesNg5YkuraREQyXTp9TmEmcJ2Z7QSuC5ZFRCSFIj0h7O4rgZXB8wPAkCjrERHJdOk0UxARkYgpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCKQ8FM/uOmZWb2TYz22JmPwrGO5nZW2a2M3jMSXVtIiKZLoqZwjHgPne/FOgPTDKzy4AHgBXu3hNYESyLiEgKpTwU3H2vu28Inh8GtgHnAyOAecFq84CbUl2biEimi/SagpnlA32BtcB57r4XYsEBfDvC0kREMlJkoWBm7YHFwBR3/6oB200ws/Vmtn7//v3JK1BEJANFEgpm1ppYILzo7q8Gw/vMrEvwehfg87q2dfcydy929+K8vLzUFCwikiGiuPvIgOeAbe7+eNxLbwDjgufjgCWprk1EJNNlRXDMAcBtwHtmtikY+wdgJrDQzO4CdgO3RlCbiEhGS3kouHsFYPW8PCSVtYiIyPH0iWYREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQkpFEREJKRQEBGRkEJBRERCUXRJzVi7H+6TuoPldEzdsUSkxdBMQUREQgoFEREJKRRERCSkUBARkZBCQUREQgoFEREJKRRERCSkzylIs/HUfb9O2bEmP/b9lB1LJJ1opiAiIiGFgoiIhBQKIiISUiiIiEhIoSAiIiHdfSQiLdaqa65N2bGufWdVyo6VTJopiIhISKEgIiIhhYKIiIQUCiIiElIoiIhISHcfiUhKDZgzIGXH+pn+xDVY2s0UzGyYme0wsw/N7IGo6xERySRpFQpm1gp4GhgOXAaMMbPLoq1KRCRzpFUoAFcCH7r7R+7+NbAAGBFxTSIiGcPcPeoaQmb2d8Awd/9vwfJtwH9x98lx60wAJgSLlwA7Ul5oy5ULVEVdhEgd9LvZtC5097y6Xki3qzBWx9hxqeXuZUBZasrJLGa23t2Lo65D5ET63UyddDt9tAf4TtxyN+CziGoREck46RYK64CeZtbdzM4GRgNvRFyTiEjGSKvTR+5+zMwmA/8PaAX80t23RFxWJtFpOUlX+t1MkbS60CwiItFKt9NHIiISIYWCiIiEFAoZxsy6mdkSM9tpZv9hZk+a2dlmVmJmS+vZZpeZ5aa6VmlZLKbCzIbHjY00s99EWZccT6GQQczMgFeB1929J3Ax0B6YEWlhkhE8dgHz74HHzSzbzNoR+92bFG1lEi+t7j6SpBsM1Lj78wDu/o2ZTQU+BsprVzKzzsB8IA/4HXV/qFCkwdz9fTP7NfBjoB3wL8BjZtYDOApMcPfNZjYdOOLuswDM7H3ghmA3y4AK4Grg98AId/+DmV0BPAdUB68Pd/feqXt3LYNmCpmlAKiMH3D3r4DdwEVxw9OACnfvS+xzIhekrELJBD8FfkCs8eVfARvdvRD4B+BXCWzfE3ja3QuAL4FbgvHngb9396uAb5q66EyhUMgsxgltQ+oZv4bY/8Hh7v8XOJj80iRTuHs18DLwAtA/eMTdfwt0NrNzT7OLj919U/C8Esg3s28BHdz934Lxl5q67kyhUMgsW4Dj+seYWUdirUX+44R19QEWSaY/Bz/19Ts7xvF/n7Ljnv8x7vk3xE6D6xRnE1EoZJYVQFszux3C7694DJhL7HxurXeAHwbrDAdyUlumZJD437USoCo4pbkL6BeM9wO6n2on7n4QOGxm/YOh0ckpt+VTKGSQ4O6PvwVuNbOdwAdADbFzufF+ClxjZhuAocSuOYgkw3Sg2Mw2AzOBccH4YqCTmW0C7ib2u3o6dwFlZraG2MzhUJNXmwHU5kJEWgQza+/uR4LnDwBd3P1HEZfV7OiWVBFpKf7GzB4k9nftE+COaMtpnjRTEBGRkK4piIhISKEgIiIhhYJIkpnZFDNrG3UdIonQNQWRJDOzXUCxu1cl8RhZ7n4sWfuXzKGZgrRYZva6mVWa2RYzmxCM3WVmH5jZSjN71syeCsbzzGyxma0LfgacYr/tzex5M3vPzDab2S3B+D+b2frgeD8NxkqBrkC5mZUHY0PNbI2ZbTCzV8ysfTD+X81se9BeenZtK3Mz6xS8l81m9q6ZFQbj082szMyWA78ys381s6K4OlfXriuSMHfXj35a5A/QKXg8B3gfOJ/YJ2U7Aa2BfwWeCtZ5CRgYPL8A2HaK/f4T8ETccs4Jx2sFrAQKg+VdQG7wPJfYp3jbBcs/Bv6RWBuHT4Huwfh8YGnwfA4wLXg+GNgUPJ9OrPfPOcHyuNq6iLVFXx/1fwP9NL8ffU5BWrJSM/vb4Pl3gNuAVe7+BYCZvULsjyfA94DLYl85AUBHM+vg7ofr2O/3iGuj4LEWCwAjgxlJFtAFuAzYfMK2/YPx1cGxzgbWAL2Aj9z942C9+cCE4PlAgk6g7v5bM4tvGveGu/8heP4K8JCZ3Q+MJ9a+RKRBFArSIgV9dL4HXOXuR81sJbADuLSeTc4K1v1DPa8ft3tOaBhoZt2B/wFc4e4HzWwuxzdxi9/2LXcfc8L2fU9zvBPVHr86HIi9z7eAEcBITmh+KJIIXVOQlupc4GDwh7IXsf9Dbwtca2Y5ZpbFX/rwAywHJtcuxJ+br8OJ6+YAHYn9gT5kZucR+66AWoeBDsHzd4EBZnZRsG1bM7sY2A70MLP8YL1RcdvX1zSuLr8AZgPramdEIg2hUJCW6jdAVtBo7RFif4x/D/wMWAu8DWzlL03TSgkas5nZVmJfG1mfR4EcM3vfzP4dGOTu/w5sJNae/JfA6rj1y4BlZlbu7vuJtV+YH9T2LtArmKHcA/zGzCqAfXG1TafupnEncfdK4CtiXzgj0mC6JVUySm3TtGCm8BrwS3d/Leq64LjaDHga2OnuP2/gProSu8jdy93/nIQypYXTTEEyzfSgHfP7xL6b+vVIqznefw9q20Ls9Nf/acjGwfdkrAX+lwJBGkszBZF6mNmdwImtl1e7+6Qo6hFJBYWCiIiEdPpIRERCCgUREQkpFEREJKRQEBGRkEJBRERCCgUREQn9fzxPHKSg8hA5AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# 1. Import libraries here\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# 2. Create the age_category column in clinical_data\n",
    "clinical_data[\"age_category\"] = np.where(clinical_data[\"Age\"]<50, \"Young\", \"Old\")\n",
    "\n",
    "# 3. Replace NaN with No cancer\n",
    "clinical_data[\"Stage\"]=np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "stage_categories=clinical_data[\"Stage\"]\n",
    "\n",
    "## this will programatically get all the stages in order\n",
    "stage_categories = np.unique(stage_categories)\n",
    "assert(np.all(stage_categories == ['No cancer', 'Stage I', 'Stage II', 'Stage III', 'Stage IV']))\n",
    "\n",
    "# 4. Loop through all the stage categories and count\n",
    "\n",
    "\n",
    "for age_cat in [\"Old\", \"Young\"]: # first get the old data, then the young data\n",
    "    print(age_cat)\n",
    "    # fill this in with the cancer category:\n",
    "    sub = clinical_data[clinical_data[\"age_category\"]==age_cat] # 1. create a subset of the data frame with the old/young patients\n",
    "    c1=0\n",
    "    c2=0\n",
    "    c3=0\n",
    "    c4=0\n",
    "    c5=0\n",
    "    for canc_cat in sub[\"Stage\"]:\n",
    "        if canc_cat==\"No cancer\":\n",
    "            c1+=1\n",
    "        if canc_cat==\"Stage I\":\n",
    "            c2+=1\n",
    "        if canc_cat==\"Stage II\":\n",
    "            c3+=1\n",
    "        if canc_cat==\"Stage III\":\n",
    "            c4+=1\n",
    "        if canc_cat==\"Stage IV\":\n",
    "            c5+=1\n",
    "    print(f\"There are {c1} patients with no cancer, {c2} patients with Stage I cancer, {c3} patients with Stage II cancer, {c4} patients with Stage III cancer, and {c5} patients with IV cancer.\")\n",
    "        # 2. count the number of patients with the stage of cancer and print\n",
    "\n",
    "\n",
    "        \n",
    "# 5. A sample barplot to compare your results\n",
    "clinical_data_copy = clinical_data\n",
    "clinical_data_copy[\"Stage\"] = np.where(pd.isna(clinical_data[\"Stage\"]), \"No cancer\", clinical_data[\"Stage\"])\n",
    "sns.countplot(\n",
    "    x = \"age_category\",\n",
    "    hue = \"Stage\",\n",
    "    hue_order = [\"No cancer\", \"Stage I\", \"Stage II\", \"Stage III\", \"Stage IV\"],\n",
    "    data = clinical_data_copy\n",
    ")\n",
    "\n",
    "plt.show() "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4747fb4c",
   "metadata": {},
   "source": [
    "## Comparing Transcriptomic and Proteomic Data\n",
    "A natural question to ask is if the levels of RNA expression correlate with protein expression. We can explore this question with the help of the `scipy` library:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "c5e207ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "from scipy import stats # we are using the stats package in particular"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1395e3d7",
   "metadata": {},
   "source": [
    "The first thing we need to do is to identify which patients and genes are shared between the transcriptomic and proteomic data sets. We need to identify which genes are shared, but luckily, we already got that data before for the patients! We will use this data to create two new data frames that contain only the rows and column names that are shared between these two data sets. \n",
    "\n",
    "**Exercise**. To make sure the data frames match:\n",
    "1. Identify the names of the genes that are shared between the two datasets (refer to the exercise when we created the dictionary). Is this data the row or column names?\n",
    "2. Create the `rna_shared` and `prot_shared` data frames as described above (Hint: how can we access rows/columns by name?).\n",
    "3. Finally, we need to make sure all the rows and columns are in the correct order."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "e2589e94",
   "metadata": {},
   "outputs": [],
   "source": [
    "# 1. Identify the genes (RNA, protein) shared between the two data sets \n",
    "shared_rna_prot = np.intersect1d(rna_data.columns, protein_data.columns)\n",
    "\n",
    "# 2. Create the two data frames\n",
    "rna_shared = rna_data.loc[name_intersects[0], shared_rna_prot]\n",
    "prot_shared = protein_data.loc[name_intersects[0], shared_rna_prot]\n",
    "\n",
    "# 3. make sure that everything is in the correct order\n",
    "assert(np.all(rna_shared.index == prot_shared.index))\n",
    "assert(np.all(rna_shared.columns == prot_shared.columns))\n",
    "\n",
    "# you can, in fact, chain comparisons in python\n",
    "# DO NOT DO THIS IN ANY OTHER LANGUAGE\n",
    "assert(rna_shared.shape == prot_shared.shape == (96, 7572))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "769b1472",
   "metadata": {},
   "source": [
    "Now, we can see how correlated the RNA and protein levels are. We'll use Spearman correlation from the `stats` library, which is `spearmanr()`. For example, to get the correlation in the KRAS gene:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "0d7aff46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The correlation is 0.078 (p = 0.461116).\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='KRAS', ylabel='KRAS'>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEGCAYAAAB7DNKzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAgbElEQVR4nO3dfZRcdZ3n8fe3m8RASCSPTUhImjjhaQQBW2BX4giIGzO4ARyRuLKMB0+W2cE4cuYsMLO7x7OsHNjZYcZWlEnAEV13AiswxD0xiGGUsALSAURCgISQQJOnTngICYYA/d0/qipUqqtu3Xq6j5/XOX2qb9Xtql/dqr7f+/v+nszdERERqaUr7gKIiEiyKVCIiEggBQoREQmkQCEiIoEUKEREJNAhcRegEyZPnuy9vb1xF0NEJDXWrFmz092nVHssk4Git7eXgYGBuIshIpIaZra51mNKPYmISCAFChERCaRAISIigRQoREQkkAKFiIgEymSvJ0mm4WFn0669bN+9j57xY+idNJauLou7WCJShwKFRGJ42Fm5dhtX3fkk+94ZZsyoLm66+BTm/eGRChYiCafUk0Ri0669B4IEwL53hrnqzifZtGtvzCUTkXoUKCQS23fvOxAkSva9M8yON/fFVCIRCUuBQiLRM34MY0Yd/HUbM6qLqePGxFQiEQlLgUIi0TtpLDddfMqBYFFqo+idNDbmkolIPWrMlkh0dRnz/vBIjl88lx1v7mPqOPV6EkkLBQqJTFeXMXvK4cyecnjcRRGRBij1JCIigRQoREQkkAKFiIgEUqAQEZFAChQiIhJIgUJERAIpUIiISKBYA4WZzTOz58xsg5ldU2OfT5rZk2a21sx+FXUZRUTyLrYBd2bWDdwMnAcMAo+Z2XJ3f6ZsnyOA7wLz3P0lM5saS2FFRHIszhrF6cAGd9/o7vuBZcCCin2+CNzt7i8BuPuOiMsoIpJ7cQaK6cDLZduDxfvKHQtMMLNfmtkaM/v3tZ7MzBaZ2YCZDQwNDXWguCIiyTQ87Gwc2sPDL+xk49Aehoe9rc8f51xP1WaDq3x3hwAfBc4FDgUeNrNH3P35EX/ovgRYAtDX19feoyQiklBRrB4ZZ41iEDi6bHsGsKXKPivdfa+77wQeBD4SUflEgM5frYm0IorVI+MMFI8Bc8zsGDMbDVwCLK/Y515grpkdYmaHAWcA6yIup+RY6Wptfv9qFi59lPn9q1m5dpuChSRGFKtHxhYo3P1d4ErgPgon/zvdfa2ZXWFmVxT3WQesBJ4CfgPc6u5Px1VmyR+t9S1JF8XqkbGuR+HuK4AVFffdUrH9N8DfRFkukZKgqzWtqyFJUFo9srKNop2rR2rhIkmN4WFn0669bN+9j57x0ayQV7paKw8WWutbkiSK1SMVKCQVoujZUU0UV2sirer06pHmnr1Gub6+Ph8YGIi7GNJGG4f2ML9/9Ygr+xWL53Y8BVSqyWitb8kyM1vj7n3VHlONQlIhzrYCrfUteafZYyUVoujZISLVKVBIKpTaCkrBQm0FItFR6klSIYqeHSJSnQKFpIbaCkTiodSTiIgEUo1CAsUxyE1EkkWBQmqKa5CbiCSLUk9SUxonxNOU4CLtpxqF1JS2CfFUAxLpDNUopKa0DXJLYw1IJA0UKKSmtA1yi2IBF5E8UupJakrbIDdNCZ4s6jGXHQoUEihNg9w0JXhyqL0oWzTNuGSKpgRPhjinhZfmaJpxyY001YCyLG095iSYGrNFpO3S1mNOgsUaKMxsnpk9Z2YbzOyagP0+ZmbvmdmfRFk+EWlO2nrMSbDYUk9m1g3cDJwHDAKPmdlyd3+myn43AvdFX0oRaUbaesxJsDjbKE4HNrj7RgAzWwYsAJ6p2O+rwF3Ax6Itnoi0Qu1F2RFn6mk68HLZ9mDxvgPMbDpwIXBLvSczs0VmNmBmA0NDQ20tqEhWaC4saUacNYpqddDKb+3fA1e7+3tmwVVWd18CLIFC99h2FFAkSzS2QZoVZ41iEDi6bHsGsKVinz5gmZltAv4E+K6ZXRBJ6UQyRnNhSbPiDBSPAXPM7BgzGw1cAiwv38Hdj3H3XnfvBX4C/Ed3/+fISyqSAZoLS5oVW+rJ3d81sysp9GbqBr7v7mvN7Iri43XbJUQkPM2FJc3SFB4iOaE2CgmiKTxEasjTDKca2yDNUqCQ3MrjFbbGNkgzNNeT5JZ6AYmEo0AhuaVeQCLhKFBIbmmGU5FwFCgktzTDqUg4asyW3FIvIJFwFCgk19QLSKQ+pZ5ERCSQAoWIiARSoBARkUAKFCIiEkiBQkREAilQiIhIIHWPFYlQnmarlexQoBCJSB5nq5VsUOpJJCKarVbSSoFCJCKarVbSSqknkTYI0/agNaslrVSjkFwbHnY2Du3h4Rd2snFoD8PDja8hX2p7mN+/moVLH2V+/2pWrt024rk0W62klbk3/o/Rthc3mwd8C+gGbnX3Gyoe/3fA1cXNPcCfuftv6z1vX1+fDwwMtLu4kjHtalzeOLSH+f2rR9QUViyeO2KywVLNQ7PVStKY2Rp376v2WGw1CjPrBm4GPgOcCCw0sxMrdnsR+CN3Pxm4DlgSbSkly5ptXK6sheza+3botofSbLVnzp7M7CmHK0hIKsTZRnE6sMHdNwKY2TJgAfBMaQd3/3XZ/o8AMyItoWRaUONyrWnHq9VCbvzcycyadCibd/3+wH5qe5AsibONYjrwctn2YPG+Wi4HflbrQTNbZGYDZjYwNDTUpiJKljWzFGq1WsjVdz3FdQtOUtuDZFacNYpqde6qDSZmdjaFQHFWrSdz9yUUU1N9fX3xNbxIapQalyvbKIJO8LVqIaO6jRVaKU8yKs5AMQgcXbY9A9hSuZOZnQzcCnzG3XdFVDbJgWaWQq3VxbVn/BitlCeZFWfq6TFgjpkdY2ajgUuA5eU7mNlM4G7gUnd/PoYySso02t210cbldnVxbUe3XJGoxFajcPd3zexK4D4K3WO/7+5rzeyK4uO3AP8VmAR818wA3q3VfUsak8XJ6aKYS6mZWkgc5RRpp1jHUXSKxlEEy+qJqpHxDHFKSzklXxI5jkLik9XJ6dIyl1JayilSokCRQ1k9UTXT3TUOaSmnSIkCRQ5l9USVlrmU0lJOkRK1UXRQUhuMs9pGAemZSykt5ZT8CGqjUKDokKSfjHWiEpFyasyOQdIbjDU5nYiEpUDRIVltMJba8jiILo/vOY+0wl2HaDWzfEl6qrET8vie80o1ig5Rz5Z8SXqqsRPy+J7zSjWKDmnHVA+S3J5jlZpZ2yLt8vie8yowUJjZLOB1d3+juH02cAGwGfiOu+/veAlTrNRgrH+a5qQptZHHVGMe33Ne1Us93QmMBTCzU4D/A7wEfAT4bkdLJrnXSGoj7kbVPKYa8/ie86pe6ulQdy+tEfElCjO8/q2ZdQFPdrRkknthUxtJqHnkMdWYx/ecV/VqFOWf+DnAKgB3H66+u0j7hJ1qJCmNqnkcm5LH95xH9QLFA2Z2p5l9C5gAPABgZtMAtU9IR4VNbWjMSmfFndaT+NVLPf0F8AVgGnCWu79TvP9I4K86WC6R0KmNTjWqpqXHVSclIa0n8QusUXjBMnf/O3d/peyhscCCzhZNJFxqoxONqqUT5Pz+1Sxc+ijz+1ezcu223F1NJyWtJ/EKPY6i2Ovpi8DFwIvAXR0qk+RUs1fwnWhUrXWCPD5nq9BprIRA/XEUxwKXAAuBXcAdFGacPTuCskmOtJriaPeYFZ0gCzRWQqB+Y/azwLnAZ939LHf/NvBe54sleZO0FEdWF3dqlMZKCNQPFJ8DtgH/YmZLzexcDu4y2xIzm2dmz5nZBjO7psrjZmb9xcefMrPT2vXakizt7rlUq6dO2B48OkEWlNJ6KxbPZdmiM1ixeK4asnMoMPXk7vcA95jZWApTd3wd6DGz7wH3uPvPm31hM+sGbgbOAwaBx8xsubs/U7bbZ4A5xZ8zgO8VbyVj2pniqJXG+vQJPfx83fZQ6S0NJnufpqKRULPHuvted/+xu58PzACeoPXusacDG9x9Y3HOqGWM7Em1APhhsffVI8ARxTEckjHtvIKvlcZau/WNhtJbGkwmUlC311Pxyn+Cu+8s3rWneHtki689HXi5bHuQkbWFavtMB7ZWKeciYBHAzJkzWyyaRK2dV/C10li79r7N5WfNxopPedeaQba+sS9XDdQaGyLNqNfr6RLgH4C9ZrYe+AbwI+AxCl1lW1Ht21mZMA6zT+FO9yXAEiismd1a0aSTap2s2pXiqJbGmjXpUN546z1ue2jjgbTT4nPmcMfAS7lpoNbgOWlWvdTTfwY+6u5HUWifWAl81d0vdPfHW3ztQeDosu0ZwJYm9pEUiWIgW7U01nULTuLae546KO3U/8B6rltwUm4aqF/cWT0l9+JODZ6TYPVST/vdfQOAuz9uZi8WG7jb4TFgjpkdA7xCYbxGZS1lOXClmS2jkJZ6w91HpJ0kPaIYyFYtjVUrHTWq21JxNd2OlNHmV/dWPQYvvbqXD03NR+pNmlMvUEw1s6vKtg8v33b3m5p9YXd/18yuBO4DuilMYb7WzK4oPn4LsAKYD2wA3gK+3OzrSTJENZCtWhqrWq+qnvHJTzu1K2U0dvQhVY/BYaO10KUEq5d6WgqMK/sp3275v9rdV7j7se7+IXf/ZvG+W4pBojTX1J8XHz/J3QdafU2JV1wD2dI8LqJdgxF7xn+Ar50756Bj8LVz59Az/gNtL7NkS71LidvcfbDaA2b22Q6URzKudMIunfhmTTqU6xacxPbd+w483olUUJrHRbSrFjZz4ljm9BzOok/MZtihy2BOz+HMnJj8YCnxqhcoVpnZv3H3TeV3mtmXKTR0/7RTBZNsKj9h7/79fl7c+RaLfjTQcEqlmZx9WgeOtWswYleXcc5xPcyefHjqgqXEq17q6evA/WY2p3SHmV0LXAX8UScLJtnV1WX0ThrL7n3vce09v2s4pZK3KcDbmTarHEQIaFEiqaveFB4rzOxt4GdmdgHwFeBjwCfc/bUIyicp0MzV/aZdexnY/GrVlMrz298ERqahSq+zaddentu2mwmHjWbrG/ua7jmVlsFnnUqbaVyFhFW3u4O7rzKzPwV+CfwaONfdtcakAM2fbLbv3sewV++J9LtXdvMXdzx50PNUe53F58zhR49sPhAsGsnZJ/Uk2enBiOW05oaEFZh6MrM3zWw38DNgPIUpx3eU3S8512yPnJ7xY/jpb19h8TkH98L5L+efyN2PD454nmqv0//Aei46bcaBv20kZ5+0ac0h+pSa1hqXsOqlnsZFVRBJp2Z75PROGsvV807gxpXruPys2XR3wfFHjueWX24A4M/P/gPMYGjP2/ROGlvzdcyay9kncWGiqK/wtSiRhKWRNtKSZk82B/LuR45jx5v7OHRUN4uXPcH+d51Lz5xF/wPr2ffOMLeu3shNF5/CcT3jqr7O3D+YzEWnTm84Z5/Ek2TUwauyq3KaxpZItEJNMy5SSys9csp74Jw0/QiunncCn++bcSBIwPtX1d1dVH2dj/VObGoK8CQOwIt6MKIWJZKwzD173eH6+vp8YECDuKNSaoBttUfOu+8Oc/+67fzZj0fON7ls0Rmc3jupLa/TbLk73UsqqQ3skg9mtsbd+6o9ptSTtKxdPXJeeu0tnt22u2ZKqN09fxp5vihO4uXdYF/d+zajurt4a/97bNq1N7FddyUfFCgi0O4r0bT0/2/U9t37uHNgkMXnzDmQfhozqovrL4x/KvBSQ/OEw0Zz0WkzMIPntu3mxGnj6J3cWOAK+vxKgxGf3famahaSGAoUHdbuK9Espyd6xo/htbf286NHNh9Yia7L4LSZR8T+3rbv3seEw0Yf1NA+ZlQXsyaNZebE8IE6zOcX5/iGrF6ESGvUmN1h7e6vn8T+/+1SamB+7a393PwvG7h19UaOP3J8U5PWDQ97qKkpwu7XM35M1Yb2v7rndw0d+zCfX1zjG/I2NYqEpxpFk8JeebW7y2MS+/+3S7NTVVR+FjMnHMbP122vW+tqpHbWO2ksx04d1/KxD/P5xdV1VyO1pRbVKJrQyJVXu7s8xrWeQ1QqJ60LEyQqP4tfb9wVqtbVSO2sq8s4Ydr4lo99mM8vTNfdsDWhRmikttSiQNGERk4w7e6vn8T+/3Gq9lnUmmyw8oTX6InxmMmtH/swn1+98Q2dShFl/SJEmqfUUxMaSf+0e+bPNC/A006ldNPz298c8VnUmmyw8oTXaIqnHcc+7HMEdd3tVIpII7WlFgWKJjRzgomr/38WlbctfGXu7BGfxU9/+wo3fu5krr7rqcATXjMnxnYc+1afo1PtVLoIkVoUKJqgK694lV9R37Vm5LiLq+edwKdP6OGk6R+se9WexhNjJxu7834RItXFMoWHmU0E7gB6gU3AxZULIZnZ0cAPgSOBYWCJu38rzPNHMYVHu6atkMY9/MJOFi599MD2tA+O4aLTZnDy9PHM6RmX+c8iy2NpJD5BU3jEFSj+B/Cqu99gZtcAE9z96op9pgHT3P1xMxsHrAEucPdn6j2/5nrKto1De5jfv3rEFfWKHHXj1IWKtFtQoIir19MC4Pbi77cDF1Tu4O5b3f3x4u9vAuuA6VEVUJKr1Z5fnehaGrVGuxGLtCKuGsXr7n5E2fZr7j4hYP9e4EHgw+5edWU9M1sELAKYOXPmRzdv3tzWMkvzGpkWIuy+zV5RK20jUl0sqScz+wWF9oVKfw3cHjZQmNnhwK+Ab7r73WFeW6mn5GjkxBzFSbxe2qpWoNIcSJJ1sUwz7u6fCijQdjOb5u5bi20RO2rsNwq4C/hx2CCRVlk9ETXS5z+KKSSCupb2ThpbNVB9+oSeUFOCiGRVXG0Uy4HLir9fBtxbuYOZGXAbsM7db4qwbJHL8mRsjYx+jmIKiaDRx9UC1Y0r1/Hk4Os8u203X5k7m2kfHJOpiRhFwogrUNwAnGdm64HzituY2VFmtqK4z8eBS4FzzOzJ4s/8eIrbWVmeEbaRaSGimEIiqCG8MlBN++AYvtA3ky/d9ij9qwqz2V565qwDwUJzIElexDLgzt13AedWuX8LML/4+0NALur1WZ4RtpHBiVEMZAxaRW7quIMHsl102shpxfsfWM/lZ83mtoc2ag4kyQ2NzE6AuKaVbkXYNpVGRj9HNVK61ipy3/niqQcFqu4uqgbw7i40El9yJZbusZ3WqV5PnWpwTluXzbSVt5pavZ9Wfm0uww473tzHoaMO4QtLHh6xzx2LzuSk6fGvuifSTrH0esqaTp4c0zbnUBYWuKmV7tu2e9+BQWzDw141FaYg0TlZ7P2XhfekQBFSp0+OaZqMLQttKmHSfWkL4GmXhZpqpay8Jy1cFJJW/3pfUha4aWUqjrDTgLRjqowsTBkShSz2/svKe1KNIqQkNjjHVaVNwjTrrV6pRVVbyMoVZRSiqKlG/T+Thdo3KFCEloSTY7k4T0BJSMm0IxUYRbovC+05Uen0xVgc/zNJvMBshlJPIdVbxzhqcVdp4569NC2pwLSUMwk6vR58HP8zWVnjXjWKBiSpwTkrVdpmNXqlFleaLitXlFHodE01jv+ZJNS+20E1ipRKSoNyJwU1Ajdypfbuu8M8tGEn//zkK/y/F3bx5R/8JrK5tLJyRdmqsA36naypxvU/E3ftux004C6lst5IGub9hVmTYnjY+elTW7j6rqcOPM/ic+Zwx8BL/OOfnh5J7Svvq9El5bualHIkVeKWQu20TgeKpAygyfIJqF3LndZ6nsvPms3cOZM4c/bktpZbRkrS0rVZ/p9plUZmt1GSrkqS1GbSbu3KJ9d6nu4uMpWmS7Iktadl+X+mk9RG0aC4exvlRbvyybWep2/WxNy1E8QlD+1pWadAURS2sS0r3R2TPlq4XY3A1Z7nxs+dzL+ePUkph4ioQT/9lHqisXRSFro7Jil9Vku7uhVmpXtimukzSD81ZtNYY1saTrL1NPp+k9BwL9Xp85F2UWN2HY00tmXh6ijs+81CUMwyfT4SFbVR0HhjW9oH0IR9v2q4TzZ9PhIVBQry19gW9v1mpeE+q/T5SFRiST2Z2UTgDqAX2ARc7O6v1di3GxgAXnH38ztRniykkxoR9v1moeEespvHz8rnI8kXV43iGmCVu88BVhW3a/kasK7TBUp7OqlRYd5vFmpapTz+/P7VLFz6KPP7V0c2z1OnZeHzkXSIpdeTmT0HfNLdt5rZNOCX7n5clf1mALcD3wSuClujyMNcT1FJ+5QHSZk+olO1mrR/PpIcSez11OPuWwGKwWJqjf3+HvhPwLh6T2hmi4BFADNnzmxTMSXtUx4kYfqITvZOSvvnI+nQsdSTmf3CzJ6u8rMg5N+fD+xw9zVh9nf3Je7e5+59U6ZMaanskh1JmD5CvZMk7ToWKNz9U+7+4So/9wLbiyknirc7qjzFx4F/a2abgGXAOWb2vzpVXsmmJOTx1TtJ0i6u1NNy4DLghuLtvZU7uPu1wLUAZvZJ4C/d/UvRFVGyIAk92tQ7SdIurl5PNwDnmdl64LziNmZ2lJmtiKlMklFx92hLQq1GpBWa60kkAq30TsrqOBBJliT2ehLJlWZ7J2k+J0kCTeEhkmDqMSVJoEAhIyR9UaM8UY8pSQKlnuQgSnUki3pMSRKoRiEHUaojWdRjSpJANQo5SBKmvJD3JWEciIgChRxEqY7k0XxOEjelnuQgSnWISCXVKOQgSnWISCUFChmh06kOjTQWSRcFComUut+KpI/aKDIm6YPl1P1WJH1Uo8iQNFytq/utSPqoRpEhabhaT8KKcyLSGAWKDEnDvEDqfiuSPko9ZUgaBsup+2001LNM2kmBIkNKV+uVbRRJu1rXSOPOSkNblaSLVrjLmFZWUpNs2Di0h/n9q0fULFcsnqvgLDVphbsc0dV69JKW5lHPMmk3BQqRFpSneSYcNprP983g2KnjOGHaeI6ZHE/AqNVWdeiobh5+YWcigpmkSyy9nsxsopndb2bri7cTaux3hJn9xMyeNbN1Zvavoi6rSJBSl+QJh43m0jNnseTBjVz5T0/wx99ezcq122IZ8FitZ9l/v+DDLF72BAuXPsr8/vjKJukUV/fYa4BV7j4HWFXcruZbwEp3Px74CLAuovKJhFJK81x02gz6H1ifiDEspZ5lKxbPZdmiM7hj0Zl8+4H1bN71+9jLJukUV6BYANxe/P124ILKHcxsPPAJ4DYAd9/v7q9HVD6RUEppHjMSNYal1FZ15uzJvLX/vQNBIgllk/SJK1D0uPtWgOLt1Cr7zAaGgH80syfM7FYzq9nP08wWmdmAmQ0MDQ11ptQiFUppnm4jsSPONRpeWtWxQGFmvzCzp6v8LAj5FIcApwHfc/dTgb3UTlHh7kvcvc/d+6ZMmdKGdyBSXynNc+Gp07n+wpMSOeJco+GlVR3r9eTun6r1mJltN7Np7r7VzKYBO6rsNggMuvujxe2fEBAoROLS1WX0Tj6cmRPHcsrRRyRuDItGw0ur4uoeuxy4DLiheHtv5Q7uvs3MXjaz49z9OeBc4JloiykSXpLHsCS5bJJ8cbVR3ACcZ2brgfOK25jZUWa2omy/rwI/NrOngFOA66MuqIhI3sVSo3D3XRRqCJX3bwHml20/CVQdUi4iItHQyGyRFiRt+g6RTlCgEGmSZmmVvNDCRSJNSsOKgiLtoEAh0qQ0rCgo0g4KFCJN0ohnyQsFCpEmacSz5IUas0WapBHPkhcKFCIt0IhnyQOlnkREJJAChYiIBFKgEBGRQAoUIiISSIFCREQCmbvHXYa2M7MhYHPc5UigycDOuAuRQDoutenYVJfF4zLL3asuD5rJQCHVmdmAu2va9go6LrXp2FSXt+Oi1JOIiARSoBARkUAKFPmyJO4CJJSOS206NtXl6riojUJERAKpRiEiIoEUKEREJJACRQaY2ffNbIeZPV1230Qzu9/M1hdvJ9T423lm9pyZbTCza6Irdee1eFw2mdnvzOxJMxuIrtTRqHFsPm9ma81s2Mxqdv3M4Xcm7HHJ7HdGgSIbfgDMq7jvGmCVu88BVhW3D2Jm3cDNwGeAE4GFZnZiZ4saqR/QxHEpc7a7n5LR/vI/YOSxeRq4CHiw1h/l9DtT97iUyeR3RoEiA9z9QeDVirsXALcXf78duKDKn54ObHD3je6+H1hW/LtMaOG4ZF61Y+Pu69z9uTp/mrvvTMjjkmkKFNnV4+5bAYq3U6vsMx14uWx7sHhfloU5LgAO/NzM1pjZoshKl3x5/M6EldnvjFa4y7dqa3aqv3TBx919i5lNBe43s2eLV5t5p+9MbZn9zqhGkV3bzWwaQPF2R5V9BoGjy7ZnAFsiKFucwhwX3H1L8XYHcA+FlIvk8zsTSpa/MwoU2bUcuKz4+2XAvVX2eQyYY2bHmNlo4JLi32VZ3eNiZmPNbFzpd+DTFBo0JZ/fmboy/51xd/2k/Af4J2Ar8A6FK77LgUkUevWsL95OLO57FLCi7G/nA88DLwB/Hfd7ScJxAWYDvy3+rM3acQk4NhcWf38b2A7cp+9MuOOS9e+MpvAQEZFASj2JiEggBQoREQmkQCEiIoEUKEREJJAChYiIBFKgEGmBme0p+31+cVbamWb2DTN7pTiT6DNmtrDi7y40Mzez48vu6zKzfjN7ujgL6WNmdkyU70ekGgUKkTYws3OBbwPz3P2l4t1/5+6nUJg07x/MbFTZnywEHqIwYK3kCxT65p/s7idR6L//eoeLLlKXAoVIi8xsLrAU+GN3f6HycXdfD7wFTCjufzjwcQqDucoDxTRgq7sPF/9u0N1f63DxRepSoBBpzQcoTANygbs/W20HMzsNWO+FOYCgMLX5Snd/Hni1+DjAncBni+mqvzWzUztcdpFQFChEWvMO8GsKtYNKXzez54BHgW+U3b+QwjoOFG8XQqEGARwHXAsMA6uKKS2RWGkKD5EWFBuzpwK/AP6vu19fvP8bwB53/59mdhGF9osPAWMpzBu0g8L03N3F21le8c9oZn9ZvP+rEb0dkaq0HoVIi9z9LTM7H1htZtvd/baKx+82s8t4f9baH7r7fyg9bma/As4ys73ANi+sadAFnAw8FdHbEKlJgUKkDdz9VTObBzxoZjur7PLfgP9NYfbR6yseuwv4IoW2jqVm9oHi/b8BvtOhIouEptSTiIgEUmO2iIgEUqAQEZFAChQiIhJIgUJERAIpUIiISCAFChERCaRAISIigf4/b+3afYBKP9cAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# we need the nan_policy=\"omit\" to throw out NaN values\n",
    "corr, pval = stats.spearmanr(rna_shared[\"KRAS\"], prot_shared[\"KRAS\"], nan_policy=\"omit\")\n",
    "\n",
    "# round() to truncate the numbers\n",
    "print(f\"The correlation is {round(corr, 3)} (p = {round(pval, 6)}).\")\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "\n",
    "sns.scatterplot(\n",
    "    x = rna_shared[\"KRAS\"],\n",
    "    y = prot_shared[\"KRAS\"],\n",
    "    ax = ax\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3174183c",
   "metadata": {},
   "source": [
    "**Exercise**. Investigate the correlation between the RNA and the protein data for the CD8A gene, printing the values. Then, using seaborn, create a scatterplot, with RNA on the x-axis and protein on the y-axis (be sure to create your fig/ax!). Does the correlation look like it makes sense? As usual, be sure to label the axes!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "a1002dc6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Patient_ID\n",
      "01CO001    7.7279\n",
      "01CO005    6.4263\n",
      "01CO006    9.1548\n",
      "01CO008    5.6439\n",
      "01CO013    6.8074\n",
      "            ...  \n",
      "21CO007    6.2288\n",
      "22CO004    7.3129\n",
      "22CO006    9.0224\n",
      "24CO005    9.0498\n",
      "27CO004    7.4429\n",
      "Name: CD8A, Length: 106, dtype: float64\n",
      "Patient_ID\n",
      "01CO005        NaN\n",
      "01CO006        NaN\n",
      "01CO008     -1.320\n",
      "01CO013        NaN\n",
      "01CO014     -0.509\n",
      "             ...  \n",
      "21CO006.N   -0.119\n",
      "21CO007.N   -0.694\n",
      "22CO004.N      NaN\n",
      "22CO006.N    1.020\n",
      "27CO004.N   -0.168\n",
      "Name: CD8A, Length: 197, dtype: float64\n",
      "Patient_ID\n",
      "01CO005    6.4263\n",
      "01CO006    9.1548\n",
      "01CO008    5.6439\n",
      "01CO013    6.8074\n",
      "01CO014    8.4009\n",
      "            ...  \n",
      "21CO006    8.3174\n",
      "21CO007    6.2288\n",
      "22CO004    7.3129\n",
      "22CO006    9.0224\n",
      "27CO004    7.4429\n",
      "Name: CD8A, Length: 96, dtype: float64\n",
      "Patient_ID\n",
      "01CO005       NaN\n",
      "01CO006       NaN\n",
      "01CO008   -1.3200\n",
      "01CO013       NaN\n",
      "01CO014   -0.5090\n",
      "            ...  \n",
      "21CO006    0.0302\n",
      "21CO007       NaN\n",
      "22CO004       NaN\n",
      "22CO006    0.5760\n",
      "27CO004       NaN\n",
      "Name: CD8A, Length: 96, dtype: float64\n",
      "The correlation is 0.802 (p = 0.0).\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:xlabel='CD8A', ylabel='CD8A'>"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAYoAAAEGCAYAAAB7DNKzAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAcU0lEQVR4nO3de5Bc5Xnn8e9vQLaMLlyk0QWJYVBZBgzEQplgCOACfInQOgZ8YcEV26FYz9qBCNZxOXa8Fbu2EhtXEnYtoMzKN8BxwPhCTNZaGQx2wFsGe8ASIMRFCAlkCWkQWFcmusyzf/SZYTTqPj3d093ndPfvU9U13afPdD89gvP0+77P+76KCMzMzErpyDoAMzPLNycKMzNL5URhZmapnCjMzCyVE4WZmaU6POsA6mH69OnR3d2ddRhmZk3jkUceeTkiOos915KJoru7m76+vqzDMDNrGpI2lHrOXU9mZpbKicLMzFI5UZiZWSonCjMzS5VZopB0nKSfS1ojabWka4qcI0lLJa2V9JikhVnEambWzrKsetoP/FVEPCppCvCIpHsj4skR51wIzE9ubwe+lvw0M7PE4GCwfttutuwYYObUiXRPm0RHh2r2+pkliojYDGxO7u+UtAaYA4xMFBcBt0VhiduHJB0laXbyu2ZmbW9wMFix+iU+dedKBvYNMnFCB9dfuoBFp8yqWbLIxRiFpG7gdODhUU/NAV4c8XhjcqzYa/RK6pPU19/fX5c4zczyZv223cNJAmBg3yCfunMl67ftrtl7ZJ4oJE0GfghcGxE7Rj9d5FeKbqAREcsioiciejo7i04uNDNrOVt2DAwniSED+wbZunOgZu+RaaKQNIFCkvhuRPyoyCkbgeNGPJ4LbGpEbGZmzWDm1IlMnHDwpXzihA5mTJlYs/fIsupJwDeBNRFxfYnT7gY+mlQ/nQls9/iEmdnruqdN4vpLFwwni6Exiu5pk2r2HllWPZ0NfAR4XNLK5NjfAF0AEXEzsBxYDKwF9gBXND5MM7P86ugQi06ZxUlLzmXrzgFmTGmtqqdfUnwMYuQ5AVzVmIjMzJpTR4eY1zmZeZ2T6/P6dXlVMzNrGU4UZmaWyonCzMxSOVGYmVkqJwozM0vlRGFmZqmcKMzMLJUThZmZpXKiMDOzVE4UZmaWyonCzMxSOVGYmVkqJwozM0vlRGFmZqmcKMzMLJUThZmZpXKiMDOzVE4UZmaWKtNEIelbkrZKeqLE8+dJ2i5pZXL720bHaGbW7jLbMztxC3AjcFvKOQ9GxHsbE46ZmY2WaYsiIh4AXskyBjMzS9cMYxRnSVol6f9KOqXUSZJ6JfVJ6uvv729kfGZmLS3vieJR4PiIeBtwA/CvpU6MiGUR0RMRPZ2dnY2Kz8ys5eU6UUTEjojYldxfDkyQND3jsMzM2kquE4WkWZKU3D+DQrzbso3KzKy9ZFr1JOl24DxguqSNwBeACQARcTPwQeCTkvYDrwGXRURkFK6ZWVvKNFFExOVlnr+RQvmsmZllJNddT2Zmlj0nCjMzS+VEYWZmqZwozMwslROFmZmlcqIwM7NUThRmZpbKicLMzFI5UZiZWSonCjMzS+VEYWZmqZwozMwslROFmZmlcqIwM7NUThRmZpbKicLMzFI5UZiZWapME4Wkb0naKumJEs9L0lJJayU9Jmlho2M0M2t3WbcobgEWpTx/ITA/ufUCX2tATGZmNkKmiSIiHgBeSTnlIuC2KHgIOErS7MZEZ2ZmkH2Lopw5wIsjHm9Mjh1CUq+kPkl9/f39DQnOzKwd5D1RqMixKHZiRCyLiJ6I6Ons7KxzWGZm7SPviWIjcNyIx3OBTRnFYmbWlvKeKO4GPppUP50JbI+IzVkHZWbWTg7P8s0l3Q6cB0yXtBH4AjABICJuBpYDi4G1wB7gimwiNbN6GxwM1m/bzZYdA8ycOpHuaZPo6CjW+2yNlmmiiIjLyzwfwFUNCsfMMjI4GKxY/RKfunMlA/sGmTihg+svXcCiU2Y5WeRA3ruezKwNrN+2ezhJAAzsG+RTd65k/bbdGUdm4ERhZjmwZcfAcJIYMrBvkK07BzKKqLzBwWBd/y5+9dzLrOvfxeBg0YLMlpBp15OZGcDMqROZOKHjoGQxcUIHM6ZMzDCq0tqtq8wtCjPLXPe0SVx/6QImTihckoYuvN3TJmUcWXHt1lXmFoWZHSSL6qOODrHolFmctORctu4cYMaUfFc9pXWVzeucnFFU9eNEYWbDsuxS6egQ8zonN8WFttm6ysbLXU9mNqzdulSq1WxdZePlFoWZDWuGLpU8TMxrtq6y8XKiMLNh4+1SqfdFPE/VRs3UVTZe7noys2Hj6VIZuogvXvogl3/9YRYvfZAVq1+q6fwCd41lwy0KMxs2ni6VUhfxk5acW/Jbd6UtkGboGmtFThRmdpBqu1QqvYhX043UbtVGeeGuJzOriaGL+EhpF/FqupHardooL9yiMLNDVDMoPXQRH91CKHURr6Ybqd2qjfLCicLMDlJtZVGlF/FKupGKJa5qxyTyUF7bbFTY8qG19PT0RF9fX9ZhmDWldf27WLz0wUMu4MtTBqWrMdaEVO68Si78eSqvzRtJj0RET7Hn3KIws4M0qrJorC2QtGqq7mmTKrrwV1OZZR7MNrNRKh2UHo+hCqsz501nXufkohf3tMRV6YB4M+57kQeZJgpJiyQ9LWmtpM8Wef48SdslrUxuf5tFnGbtJG+VRWmJq9ILfyOTYCvJrOtJ0mHATcC7gY3AbyTdHRFPjjr1wYh4b8MDNGtTeassKlZN9aVLTqNDMGNKZfMqKq3MsoLMBrMlnQV8MSL+JHn8OYCI+PKIc84DPl1povBgtllrGRwMnn95N2te2sEzW3by/b6NvLpnLzd++HT27o+KBqeHBr/zkATzJK+D2XOAF0c83gi8vch5Z0laBWyikDRWF3sxSb1AL0BXV1eNQzWzLHV0CAk+/f1VB7Uerv6X37LimnNZXkHrp50W86uVLBNFsX/J0c2bR4HjI2KXpMXAvwLzi71YRCwDlkGhRVHDOM0sB0qNR7y0Y2B4MNzqI8vB7I3AcSMez6XQahgWETsiYldyfzkwQdL0xoVoZnnhgejsZJkofgPMl3SCpDcAlwF3jzxB0ixJSu6fQSHebQ2P1MxqZnAwWNe/i1899zLr+neNeRnyvFVjtZPMup4iYr+kq4GfAocB34qI1ZI+kTx/M/BB4JOS9gOvAZdFK04lN2sT45kZXY9qLC/nMTZewsPMGqZRy4OMhZfzOFha1ZNnZptZw+RpZrR3yxs7Jwoza5jxDkhXO75RTJ6SVt45UZjZQWp5MR4tT3tyu4pq7Koao5B0NvDhiLiq9iGNn8cozKrTiH77amdG13p8w2MUB6vJzGxJC4APA5cCzwM/qkl0ZpYbjViGu1F7co8ljjytaZVnqYlC0lsozG+4nML8he9RaIWc34DYzKzBGrUXRTUq2RFvrLycx9iUG6N4Cngn8KcRcU5E3AAcqH9YZlYLlY435Lnf3hPuslOu6+kDFFoUP5e0AriD4ms0mVnOVNMHn+dluN1VlJ0xDWZLmgRcTKEL6gLgVuCuiLinrtFVyYPZZtUP/noZ7vY07gl3EbE7Ir6b7AsxF1gJHLIjnZnlR7XzBMayPam1l7KJQtLhI+5PBuYB34+IC+oZmJmNT57HG6y5pCYKSX8ObJH0jKQLgceArwCrJF3egPjMrEoe/LVaKTeY/VfAicAUYBVwekQ8J2kmcC9we53jM7MqefDXaqVcojgQES8DL0vaFRHPAUTElmSbCDPLMc8TsFoolyhekPRlCi2KpyT9E4UZ2e8CNtc7ODMzy165wew/A3ZQ2Lb0fcCvgM8BM4E/r2tkZmaWC6ktiojYAXx5xKEfJDczawPeAc5gbOWxH5P0qKTdya1P0kdr8eaSFkl6WtJaSYfMy1DB0uT5xyQtrMX7mll5tV7W25pXufLYjwLXUqh+OhaYA3wGuGa8yULSYcBNwIXAW4HLJb111GkXAvOTWy/wtfG8p5mNnXeAsyHlWhR/AVwSET+PiO0R8fuIuJ/CGlB/Mc73PgNYGxHrImIvhXWkLhp1zkXAbVHwEHCUpNnjfF+zplDPDYTGwjvA2ZByVU9TI2L96IMRsV7S1HG+9xzgxRGPNwJvH8M5cyhScSWpl0Krg66urnGGZpatahb0q/V4Qj2W9bbmVK5F8VqVz41Fsf+CR39lGss5hYMRyyKiJyJ6Ojs7xxmaWbYq7fYpNZ6wf/9g1a2SrqOP4O8uPvWgmd1/d/GpdB19xPg/oDWVci2KkyU9VuS4KKz5NB4bgeNGPJ4LbKriHLOWU+kGQqUSy7KP9ND7nb6qtvp84dU93HD/s1x5zjwkiIAb7n+WhV1HewJfmymXKN5GYc7Ei6OOH8/4L9i/AeZLOgH4HYV9Lz486py7gasl3UGhW2p7RHiin7W8Srt9SiWWvg2vHJI85vSeyZ69B8p2T23ZMcCGba9x08/XHnQ8D7vdWWOV63r6n8COiNgw8gbsSZ6rWkTsB64GfgqsAe6MiNWSPiHpE8lpy4F1wFrg64x/AN2sKVS6oF+plWIPHJw7GNg3yH1PbR1TuatXn7UhqRsXSXoiIk4t8dzjEXFa3SIbB29cZK2gkg2Eig1+f+mS0/jqfc+wYdvrw4kTJ3Rw5TnzhlsJaRsZVTOgbs0rbeOicl1PaV8d3lR9SGZWTiUL+g2tFHviX57Lmpd28MyWndzy/57nsj/q4qv3PTt8oV9ywXy+89CG4d8rNe4xlKSOPmIC3+s9i30HDnDMpDd6ZnabKpcofiPp4xHx9ZEHJV0JPFK/sMysUh0dQoJPf3/V8LhE/6699L5jHqfNOZL/2DfIP9zzFJu3vz4PolhXUqmWxMKuY5wk2lS5MYprgSsk/ULSPyW3fwf+C3BN3aMzs4qMHtTevH2Apfet5cBg8PfL1/Cfe7oOGvf40iWnHTLu4RnZNlq5RQG3AH8s6XxgaKziJ8nsbDPLmVLVUrOPnMire/bynYc2DJe7dggWdh11SCuh0tJca31lFwUESJbwuCG5OUmY5VSpaqlTZh/J9Zcu4NU9e7np52v5xoPrOGnWVLqOObSKytVONlpq1VOzctWTtbNS1VJjraJytVN7Sqt6cqIwawG1XuepktJcaw3jKY81s5yrVQugWLLxmISBE4VZXTRyZ7hSVUonlZhIVypedzdZKWMazDazsWv0znC12DfCJbGWxonCrMYafdGtRZWSNymyNE4UZjXW6ItupQsIFuOSWEvjMQqzGmv0znBD6zydtOTcqquUhpLN6DGK7mmTGjreYvnk8lizGmvWgeFiJbFAU34Wq5znUZg1WKXzEPL6rX1d/y4WL33wkNZRqaXJrXl5HoVZg1WyRHieWyBe98nAg9lmqQYHg3X9u/jVcy+zrn9XXUpc81ya6kFuAycKs2Gjk8L+/YMNmQ+R59LUWlRUWfPLpOtJ0jHA94BuYD1waUS8WuS89cBO4ACwv1T/mdl4Fev+WfaRnqLf9Of0nsmevQdqNpbQ6CqpSoyuqOqcPJHDOuDh57flaizF6iurFsVngfsiYj5wX/K4lPMjYoGThNVTse6fvg2vFP2mf99TW2vawsj7t/ah8ZYzuqfx9JadLPpqY2acW35klSguAm5N7t8KXJxRHGZA8e6fwaBo//yB5LRajSUMfWtfvuRc7uh9O8uXnJuLgezR8jyWYvWVVaKYGRGbAZKfM0qcF8A9kh6R1Jv2gpJ6JfVJ6uvv769xuNbqig3a/tuq3/GVD/zBQd/0l1wwnx89unH4nFqNJQx9az9z3nTmdU7OXZIA2Lb7P7jynHlcfcGbufqCNzP7yIklP38jigCsceo2RiHpZ8CsIk99voKXOTsiNkmaAdwr6amIeKDYiRGxDFgGhXkUFQdsba3YzOS/XnQy7zl5JqfNOZKtOwd404TDWHLHb9m8/fULY17GEuptcDDY9PsBvvnLdcN/nyUXzOd7fS8c8vnzXO5r1clkwp2kp4HzImKzpNnALyLixDK/80VgV0T8Y7nX94Q7q0a5neG27Bhg/4Hgv//4cTZse62tLoClJt4t+0gP57x5+kGf35P0mlMeJ9zdDXwMuC75+ePRJ0iaBHRExM7k/nuA/9HQKK2tFJskV+zb8Vc+8AfMOWoix0x6Y9tU/ZQq4Z1wmA75/J6k13qyGqO4Dni3pGeBdyePkXSspOXJOTOBX0paBfwa+ElErMgkWmtbxQZw//qHj3HMpDfmdiyhHkpNvJs59dBuN0/Saz2ZJIqI2BYR74yI+cnPV5LjmyJicXJ/XUS8LbmdEhF/n0Ws1t7yPBmukSop4c17ua9Vzms9maXI82S4RqpkKfNaLHtu+eIlPMxS+Nvx6yop4W2Gcl8bO7cozFL427GZE4VZWZUsGW7Witz1ZGZmqdyisFzJ605vZu3MicJyw0s/mOWTu54sN7w6qVk+OVFYbnhym1k+OVFYbnjpB7N8cqKw3PDkNrN88mC25YYnt5nlkxOF5Yont5nlj7uezMwslVsU1vY8yc8snROFtTVP8jMrz11P1tY8yc+svEwShaQPSVotaVBS0c28k/MWSXpa0lpJn21kjNYePMnPrLysWhRPAO8HHih1gqTDgJuAC4G3ApdLemtjwrN24Ul+ZuVltWf2moh4usxpZwBrk72z9wJ3ABfVPzprJ57kZ1Zengez5wAvjni8EXh7qZMl9QK9AF1dXfWNzFqGJ/mZlVe3RCHpZ8CsIk99PiJ+PJaXKHIsSp0cEcuAZQA9PT0lzzMbzZP8zNLVLVFExLvG+RIbgeNGPJ4LbBrnazY11/ubWRby3PX0G2C+pBOA3wGXAR/ONqTsuN7fzLKSVXnsJZI2AmcBP5H00+T4sZKWA0TEfuBq4KfAGuDOiFidRbx54Hp/M8tKJi2KiLgLuKvI8U3A4hGPlwPLGxhabqXV+7tv3czqyTOzm4Tr/c0sK04UTcL1/maWlTwPZtsIrvc3s6w4UTQR1/tbMS6btnpzojBrYi6btkbwGIVZE3PZtDWCE4VZE/My6dYIThRmTcxl09YIThRW1OBgsK5/F7967mXW9e9icNDrLOaRy6atETyYbYfwAGnzcNm0NYIThR1iaID06CPewPsXzkWCp1/awVtnT6F7uktz88Zl01ZvThR2iC07Bjj6iDfwkTOPZ+n9zw63Ko6fNomuY/xt1azdeIwi4T75182cOpEP9cwdThJQqKT5m7sed9mlWRtyiwL3yY/WPW0Sb5kxxavVmhngFgVQ+0lLzd466egQJ8+e6rJLMwOcKIDaTloaap0sXvogl3/9YRYvfZAVq19qumRxwnSXXZpZgbueeH3S0shkUe2351Ktk5OWnNtUXTYuuzSzIW5RUNtJS620pMJQ2eWZ86Yzr3Oyk4RZm8qkRSHpQ8AXgZOBMyKir8R564GdwAFgf0T01COeWn57rmXrxMwsD7JqUTwBvB94YAznnh8RC+qVJIbU6tuzl1Qws1aTSYsiItYASK3XleG+fTNrNXkfzA7gHkkB/O+IWFbqREm9QC9AV1dXg8IrzksqmFkrqVuikPQzYFaRpz4fET8e48ucHRGbJM0A7pX0VEQU7a5KksgygJ6enuaqRTUzy7G6JYqIeFcNXmNT8nOrpLuAMxjbuIaZmdVIbstjJU2SNGXoPvAeCoPgZmbWQJkkCkmXSNoInAX8RNJPk+PHSlqenDYT+KWkVcCvgZ9ExIos4jUza2dZVT3dBdxV5PgmYHFyfx3wtgaHZmZmo+S268nMzPLBicLMzFI5UZiZWSonCjMzS5X3mdlmLWNwMFi/bTdbdgwwc6qXdrHm4URh1gDebteambuezBqg1tvtmjWSE4VZA7TShlbWfpwozBpgaEOrkbyhlTULJwqzBvCGVtbMPJht1gDe0MqamROFWYN4QytrVu56MjOzVE4UZmaWyonCzMxSOVGYmVkqJwozM0uliMg6hpqT1A9sSB5OB17OMJzxavb4wZ8hD5o9fvBnqLfjI6Kz2BMtmShGktQXET1Zx1GtZo8f/BnyoNnjB3+GLLnryczMUjlRmJlZqnZIFMuyDmCcmj1+8GfIg2aPH/wZMtPyYxRmZjY+7dCiMDOzcXCiMDOzVC2dKCQdJum3kv5P1rFUQ9J6SY9LWimpL+t4qiHpKEk/kPSUpDWSzso6prGSdGLytx+67ZB0bdZxVUrSf5O0WtITkm6X1FS7JUm6Jol9dbP8/SV9S9JWSU+MOHaMpHslPZv8PDrLGCvR0okCuAZYk3UQ43R+RCxoxtrrxFeBFRFxEvA2mujfIyKeTv72C4A/BPYAd2UbVWUkzQGWAD0RcSpwGHBZtlGNnaRTgY8DZ1D47+e9kuZnG9WY3AIsGnXss8B9ETEfuC953BRaNlFImgv8J+AbWcfSriRNBd4BfBMgIvZGxO8zDap67wSei4gNZc/Mn8OBN0k6HDgC2JRxPJU4GXgoIvZExH7g34FLMo6prIh4AHhl1OGLgFuT+7cCFzcypvFo2UQB/C/gM8BgmfPyLIB7JD0iqTfrYKowD+gHvp10AX5DUrPu/XkZcHvWQVQqIn4H/CPwArAZ2B4R92QbVUWeAN4haZqkI4DFwHEZx1StmRGxGSD5OSPjeMasJROFpPcCWyPikaxjGaezI2IhcCFwlaR3ZB1QhQ4HFgJfi4jTgd00UXN7iKQ3AO8Dvp91LJVK+sEvAk4AjgUmSfqzbKMau4hYA3wFuBdYAawC9mcaVBtqyUQBnA28T9J64A7gAkn/nG1IlYuITcnPrRT6xs/INqKKbQQ2RsTDyeMfUEgczeZC4NGI2JJ1IFV4F/B8RPRHxD7gR8AfZxxTRSLimxGxMCLeQaE759msY6rSFkmzAZKfWzOOZ8xaMlFExOciYm5EdFPoMrg/IprmWxSApEmSpgzdB95DoRneNCLiJeBFSScmh94JPJlhSNW6nCbsdkq8AJwp6QhJovBv0DQFBQCSZiQ/u4D307z/FncDH0vufwz4cYaxVOTwrAOwkmYCdxX+3+Zw4F8iYkW2IVXlL4HvJt0364ArMo6nIkm/+LuB/5p1LNWIiIcl/QB4lEKXzW9pvmUkfihpGrAPuCoiXs06oHIk3Q6cB0yXtBH4AnAdcKekKykk8A9lF2FlvISHmZmlasmuJzMzqx0nCjMzS+VEYWZmqZwozMwslROFmZmlcqIwGwdJsyTdIek5SU9KWi7pLZJeS5YtWSPp15I+NuJ3jpT0b5JWJSuiXjHqNS+RFJJOavwnMjuU51GYVSmZwHYXcGtEXJYcW0BhDsxzybIlSJoH/EhSR0R8G7gKeDIi/lRSJ/C0pO9GxN7kpS8HfklhsugXG/mZzIpxi8KseucD+yLi5qEDEbESeHHkSRGxDvgUheW+obDY45Qk0UymsCzFfgBJkyksQXMlTbQcuLU2Jwqz6p0KjHXhyUeBoa6kGyksn70JeBy4JiKGVjm+mML+Hc8Ar0hqxrWxrMU4UZg1hkbc/xNgJYXVXBcANyZ7d0Ch2+mO5P4dyWOzTHmMwqx6q4EPjvHc03l9Mb4rgOuisH7OWknPAydJeg64ADhVUlDYjS4kfSa81o5lyC0Ks+rdD7xR0seHDkj6I+D4kSdJ6qawedANyaEXKKziiqSZwIkUFkz8IHBbRBwfEd0RcRzwPHBOnT+HWSovCmg2DpKOpbCb4h8CA8B64FrgMeApYCKwk8LmTd8e8Tu3ALMpdEldFxH/LOkXyf0VI15/CXByRHyyIR/IrAgnCjMzS+WuJzMzS+VEYWZmqZwozMwslROFmZmlcqIwM7NUThRmZpbKicLMzFL9fybL51rWIFVrAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "gene = \"CD8A\" # this will save you some typing and make the code more flexible for exploring other genes\n",
    "print(rna_data[gene])\n",
    "print(protein_data[gene])\n",
    "print(rna_shared[gene])\n",
    "print(prot_shared[gene])\n",
    "# get correlation and print\n",
    "\n",
    "corr, pval = stats.spearmanr(rna_shared[gene], prot_shared[gene], nan_policy = \"omit\")\n",
    "\n",
    "# create scatter plot\n",
    "\n",
    "print(f\"The correlation is {round(corr, 3)} (p = {round(pval, 6)}).\")\n",
    "\n",
    "fig, ax = plt.subplots(1,1)\n",
    "\n",
    "sns.scatterplot(\n",
    "    x = rna_shared[gene],\n",
    "    y = prot_shared[gene],\n",
    "    ax = ax\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "556dffec",
   "metadata": {},
   "source": [
    "## Heatmaps\n",
    "A problem arises if we want to compare many correlations. Heatmaps are useful for visualizing a large number of comparisons. To make a heatmap, we'll use the `heatmap()` function from seaborn (the `cluster()` function is closely related; the difference is that `clustermap()` will reorganize the rows/columns so that similar rows/columns are grouped near each other). Let's do the following: \n",
    "\n",
    "1. Set up our data frame to hold all comparisons. All you need to do is to access the first 20 gene names.\n",
    "2. Calculate the correlations for the first 20 genes (just to save time). You'll need to use two for loops.\n",
    "3. Call `clustermap()` -- this is filled in for you!\n",
    "4. Interpret the data. There is a \"light\" diagonal along the heatmap from the top-left to bottom-right. Is this expected? Why or why not?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "id": "a95838ba",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['01CO001', '01CO005', '01CO006', '01CO008', '01CO013', '01CO014',\n",
      "       '01CO015', '01CO019', '01CO022', '05CO002', '05CO003', '05CO005',\n",
      "       '05CO006', '05CO007', '05CO011', '05CO014', '05CO015', '05CO020',\n",
      "       '05CO026'],\n",
      "      dtype='object', name='Patient_ID')\n"
     ]
    },
    {
     "ename": "ValueError",
     "evalue": "Shape of passed values is (20, 20), indices imply (19, 19)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[0;32m/var/folders/v5/0w_yh6455q93m3x99xvqg5b80000gn/T/ipykernel_15344/1406231198.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;31m# It's good practice to declare your data frame beforehand (it's much faster than appending to a list)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0;31m# We fill everything in with 0 just as a placeholder\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m corr_df = pd.DataFrame(np.ndarray(shape=(ncomparisons, ncomparisons), dtype=np.float16),\n\u001b[0m\u001b[1;32m     10\u001b[0m                       \u001b[0mindex\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mgene_names\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m                       columns = gene_names)\n",
      "\u001b[0;32m~/opt/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__init__\u001b[0;34m(self, data, index, columns, dtype, copy)\u001b[0m\n\u001b[1;32m    692\u001b[0m                 )\n\u001b[1;32m    693\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 694\u001b[0;31m                 mgr = ndarray_to_mgr(\n\u001b[0m\u001b[1;32m    695\u001b[0m                     \u001b[0mdata\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    696\u001b[0m                     \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36mndarray_to_mgr\u001b[0;34m(values, index, columns, dtype, copy, typ)\u001b[0m\n\u001b[1;32m    349\u001b[0m     )\n\u001b[1;32m    350\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 351\u001b[0;31m     \u001b[0m_check_values_indices_shape_match\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindex\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    352\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    353\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtyp\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"array\"\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/miniconda3/envs/qbio_student/lib/python3.9/site-packages/pandas/core/internals/construction.py\u001b[0m in \u001b[0;36m_check_values_indices_shape_match\u001b[0;34m(values, index, columns)\u001b[0m\n\u001b[1;32m    420\u001b[0m         \u001b[0mpassed\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshape\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    421\u001b[0m         \u001b[0mimplied\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 422\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mValueError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"Shape of passed values is {passed}, indices imply {implied}\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    423\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    424\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mValueError\u001b[0m: Shape of passed values is (20, 20), indices imply (19, 19)"
     ]
    }
   ],
   "source": [
    "ncomparisons = 20 # define this variable in case we want to change the number of correlations to test\n",
    "                  # this makes it less likely you'll forget to change a number, e.g. in the data frame shape\n",
    "gene_names = rna_data.index[0:19] # get the first ncomparisons gene names\n",
    "print(gene_names)\n",
    "\n",
    "# Don't worry about this code\n",
    "# It's good practice to declare your data frame beforehand (it's much faster than appending to a list)\n",
    "# We fill everything in with 0 just as a placeholder\n",
    "corr_df = pd.DataFrame(np.ndarray(shape=(ncomparisons, ncomparisons), dtype=np.float16),\n",
    "                      index = gene_names,\n",
    "                      columns = gene_names)\n",
    "# This line of code didn't work for some reason, so I could not continue this exercise :'('\n",
    "\n",
    "# 2. fill in the data frame!\n",
    "for g1 in gene_names:\n",
    "    for g2 in gene_names:\n",
    "        corr, pval = stats.spearmanr(rna_shared[g2], prot_shared[g2], nan_policy = \"omit\")\n",
    "        # calculate the correlations between protein and RNA\n",
    "        corr_df.loc[g1,g2] = corr\n",
    "        # then, use .loc[] to store the correlation in corr_df\n",
    "\n",
    "# 3. create the heat map\n",
    "sns.heatmap(\n",
    "    corr_df,\n",
    "    cmap='mako'\n",
    ")\n",
    "\n",
    "plt.show()\n",
    "\n",
    "# 4. interpret!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "attached-charity",
   "metadata": {},
   "source": [
    "### Saving pandas dataframes\n",
    "#While you can certainly reload in the data every time you wish to use it, we also want to make sure you know how to save and read in a pandas dataframes. We will utilize the `to_csv()` function and `read_csv()` function. Notice in the `read_csv()` function we utilize the `index_col` argument. To observe what this argument does, try removing it and looking at the dataframe. Notice what the index values are. \n",
    "\n",
    "**Exercise**. To save our data:\n",
    "1. Make a folder in your `analysis_data` folder called `cptac`.\n",
    "2. Edit the path to save into `analysis_data`\n",
    "3. Also save the RNA and clinical data as well to the same directory (with the same name convention)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "resistant-approach",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>A1CF</th>\n",
       "      <th>A2M</th>\n",
       "      <th>A4GALT</th>\n",
       "      <th>AAAS</th>\n",
       "      <th>AACS</th>\n",
       "      <th>AADACP1</th>\n",
       "      <th>AADAT</th>\n",
       "      <th>AAED1</th>\n",
       "      <th>AAGAB</th>\n",
       "      <th>AAK1</th>\n",
       "      <th>...</th>\n",
       "      <th>ZW10</th>\n",
       "      <th>ZWILCH</th>\n",
       "      <th>ZWINT</th>\n",
       "      <th>ZXDA</th>\n",
       "      <th>ZXDB</th>\n",
       "      <th>ZXDC</th>\n",
       "      <th>ZYG11B</th>\n",
       "      <th>ZYX</th>\n",
       "      <th>ZZEF1</th>\n",
       "      <th>ZZZ3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Patient_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>01CO001</th>\n",
       "      <td>10.1984</td>\n",
       "      <td>13.8117</td>\n",
       "      <td>7.0224</td>\n",
       "      <td>9.4115</td>\n",
       "      <td>9.7432</td>\n",
       "      <td>4.5850</td>\n",
       "      <td>8.4635</td>\n",
       "      <td>7.6221</td>\n",
       "      <td>11.2336</td>\n",
       "      <td>8.3219</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1046</td>\n",
       "      <td>10.5765</td>\n",
       "      <td>10.5294</td>\n",
       "      <td>7.0553</td>\n",
       "      <td>8.4263</td>\n",
       "      <td>8.6865</td>\n",
       "      <td>10.0307</td>\n",
       "      <td>9.1699</td>\n",
       "      <td>10.2784</td>\n",
       "      <td>10.4263</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO005</th>\n",
       "      <td>10.1586</td>\n",
       "      <td>12.6434</td>\n",
       "      <td>6.7279</td>\n",
       "      <td>8.9972</td>\n",
       "      <td>9.9084</td>\n",
       "      <td>4.7004</td>\n",
       "      <td>8.2046</td>\n",
       "      <td>7.0768</td>\n",
       "      <td>10.5362</td>\n",
       "      <td>9.6402</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6883</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>9.2808</td>\n",
       "      <td>7.1799</td>\n",
       "      <td>9.2946</td>\n",
       "      <td>9.1799</td>\n",
       "      <td>9.8408</td>\n",
       "      <td>10.5622</td>\n",
       "      <td>10.4512</td>\n",
       "      <td>10.4305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO006</th>\n",
       "      <td>9.2119</td>\n",
       "      <td>15.0040</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>9.7056</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>8.4676</td>\n",
       "      <td>7.1293</td>\n",
       "      <td>10.9337</td>\n",
       "      <td>9.5661</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8424</td>\n",
       "      <td>9.8611</td>\n",
       "      <td>10.2107</td>\n",
       "      <td>7.6582</td>\n",
       "      <td>8.7415</td>\n",
       "      <td>9.0741</td>\n",
       "      <td>10.3309</td>\n",
       "      <td>10.0512</td>\n",
       "      <td>10.2703</td>\n",
       "      <td>10.5430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO008</th>\n",
       "      <td>10.7322</td>\n",
       "      <td>13.9284</td>\n",
       "      <td>4.9542</td>\n",
       "      <td>9.1497</td>\n",
       "      <td>8.9425</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.4179</td>\n",
       "      <td>7.7074</td>\n",
       "      <td>10.7194</td>\n",
       "      <td>9.3376</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6812</td>\n",
       "      <td>9.0795</td>\n",
       "      <td>9.5118</td>\n",
       "      <td>8.9159</td>\n",
       "      <td>9.9542</td>\n",
       "      <td>10.0168</td>\n",
       "      <td>10.3365</td>\n",
       "      <td>9.9307</td>\n",
       "      <td>10.2538</td>\n",
       "      <td>10.8925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>01CO013</th>\n",
       "      <td>9.7160</td>\n",
       "      <td>12.9963</td>\n",
       "      <td>6.3576</td>\n",
       "      <td>9.6202</td>\n",
       "      <td>9.2738</td>\n",
       "      <td>4.1699</td>\n",
       "      <td>7.0980</td>\n",
       "      <td>5.9069</td>\n",
       "      <td>10.2252</td>\n",
       "      <td>9.7944</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1344</td>\n",
       "      <td>9.3174</td>\n",
       "      <td>10.0238</td>\n",
       "      <td>7.4429</td>\n",
       "      <td>8.7381</td>\n",
       "      <td>9.6165</td>\n",
       "      <td>10.3848</td>\n",
       "      <td>11.3066</td>\n",
       "      <td>11.2174</td>\n",
       "      <td>10.7616</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21CO007</th>\n",
       "      <td>10.0768</td>\n",
       "      <td>13.4762</td>\n",
       "      <td>6.4594</td>\n",
       "      <td>9.6018</td>\n",
       "      <td>9.1573</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>6.2095</td>\n",
       "      <td>10.9159</td>\n",
       "      <td>10.9315</td>\n",
       "      <td>...</td>\n",
       "      <td>10.1293</td>\n",
       "      <td>9.5622</td>\n",
       "      <td>10.2958</td>\n",
       "      <td>7.5314</td>\n",
       "      <td>8.3707</td>\n",
       "      <td>8.8202</td>\n",
       "      <td>9.8234</td>\n",
       "      <td>9.4136</td>\n",
       "      <td>9.7245</td>\n",
       "      <td>10.6777</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO004</th>\n",
       "      <td>10.9293</td>\n",
       "      <td>12.6131</td>\n",
       "      <td>5.5236</td>\n",
       "      <td>9.3332</td>\n",
       "      <td>9.2877</td>\n",
       "      <td>5.0000</td>\n",
       "      <td>7.3663</td>\n",
       "      <td>6.8580</td>\n",
       "      <td>10.9031</td>\n",
       "      <td>8.8361</td>\n",
       "      <td>...</td>\n",
       "      <td>9.9687</td>\n",
       "      <td>9.5736</td>\n",
       "      <td>10.5108</td>\n",
       "      <td>8.4346</td>\n",
       "      <td>9.0768</td>\n",
       "      <td>8.8857</td>\n",
       "      <td>10.2715</td>\n",
       "      <td>9.4737</td>\n",
       "      <td>9.6511</td>\n",
       "      <td>10.9114</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22CO006</th>\n",
       "      <td>9.2432</td>\n",
       "      <td>14.2620</td>\n",
       "      <td>7.5546</td>\n",
       "      <td>9.3264</td>\n",
       "      <td>8.7039</td>\n",
       "      <td>5.5546</td>\n",
       "      <td>7.3129</td>\n",
       "      <td>8.0389</td>\n",
       "      <td>11.1592</td>\n",
       "      <td>9.2597</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8903</td>\n",
       "      <td>9.9972</td>\n",
       "      <td>10.1923</td>\n",
       "      <td>7.4998</td>\n",
       "      <td>8.1447</td>\n",
       "      <td>8.7582</td>\n",
       "      <td>10.1497</td>\n",
       "      <td>10.4062</td>\n",
       "      <td>10.2491</td>\n",
       "      <td>10.3902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24CO005</th>\n",
       "      <td>8.6546</td>\n",
       "      <td>13.0250</td>\n",
       "      <td>8.0498</td>\n",
       "      <td>10.2228</td>\n",
       "      <td>9.3729</td>\n",
       "      <td>3.1699</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>8.3399</td>\n",
       "      <td>11.0566</td>\n",
       "      <td>8.0980</td>\n",
       "      <td>...</td>\n",
       "      <td>9.8933</td>\n",
       "      <td>10.2621</td>\n",
       "      <td>10.8727</td>\n",
       "      <td>6.2479</td>\n",
       "      <td>7.8392</td>\n",
       "      <td>8.1849</td>\n",
       "      <td>9.6917</td>\n",
       "      <td>8.3264</td>\n",
       "      <td>7.8704</td>\n",
       "      <td>10.3740</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27CO004</th>\n",
       "      <td>9.8025</td>\n",
       "      <td>13.1139</td>\n",
       "      <td>7.9069</td>\n",
       "      <td>9.7582</td>\n",
       "      <td>9.2479</td>\n",
       "      <td>6.0224</td>\n",
       "      <td>7.7682</td>\n",
       "      <td>7.0444</td>\n",
       "      <td>11.0049</td>\n",
       "      <td>9.1997</td>\n",
       "      <td>...</td>\n",
       "      <td>9.6439</td>\n",
       "      <td>9.8857</td>\n",
       "      <td>10.3049</td>\n",
       "      <td>7.6865</td>\n",
       "      <td>8.6110</td>\n",
       "      <td>9.1624</td>\n",
       "      <td>9.3106</td>\n",
       "      <td>11.0532</td>\n",
       "      <td>9.6777</td>\n",
       "      <td>9.6900</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>106 rows × 13482 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "               A1CF      A2M  A4GALT     AAAS    AACS  AADACP1   AADAT  \\\n",
       "Patient_ID                                                               \n",
       "01CO001     10.1984  13.8117  7.0224   9.4115  9.7432   4.5850  8.4635   \n",
       "01CO005     10.1586  12.6434  6.7279   8.9972  9.9084   4.7004  8.2046   \n",
       "01CO006      9.2119  15.0040  6.8580   9.7056  9.1997   5.0000  8.4676   \n",
       "01CO008     10.7322  13.9284  4.9542   9.1497  8.9425   6.0224  7.4179   \n",
       "01CO013      9.7160  12.9963  6.3576   9.6202  9.2738   4.1699  7.0980   \n",
       "...             ...      ...     ...      ...     ...      ...     ...   \n",
       "21CO007     10.0768  13.4762  6.4594   9.6018  9.1573   5.0000  7.6865   \n",
       "22CO004     10.9293  12.6131  5.5236   9.3332  9.2877   5.0000  7.3663   \n",
       "22CO006      9.2432  14.2620  7.5546   9.3264  8.7039   5.5546  7.3129   \n",
       "24CO005      8.6546  13.0250  8.0498  10.2228  9.3729   3.1699  7.0444   \n",
       "27CO004      9.8025  13.1139  7.9069   9.7582  9.2479   6.0224  7.7682   \n",
       "\n",
       "             AAED1    AAGAB     AAK1  ...     ZW10   ZWILCH    ZWINT    ZXDA  \\\n",
       "Patient_ID                            ...                                      \n",
       "01CO001     7.6221  11.2336   8.3219  ...  10.1046  10.5765  10.5294  7.0553   \n",
       "01CO005     7.0768  10.5362   9.6402  ...   9.6883   9.3264   9.2808  7.1799   \n",
       "01CO006     7.1293  10.9337   9.5661  ...   9.8424   9.8611  10.2107  7.6582   \n",
       "01CO008     7.7074  10.7194   9.3376  ...   9.6812   9.0795   9.5118  8.9159   \n",
       "01CO013     5.9069  10.2252   9.7944  ...  10.1344   9.3174  10.0238  7.4429   \n",
       "...            ...      ...      ...  ...      ...      ...      ...     ...   \n",
       "21CO007     6.2095  10.9159  10.9315  ...  10.1293   9.5622  10.2958  7.5314   \n",
       "22CO004     6.8580  10.9031   8.8361  ...   9.9687   9.5736  10.5108  8.4346   \n",
       "22CO006     8.0389  11.1592   9.2597  ...   9.8903   9.9972  10.1923  7.4998   \n",
       "24CO005     8.3399  11.0566   8.0980  ...   9.8933  10.2621  10.8727  6.2479   \n",
       "27CO004     7.0444  11.0049   9.1997  ...   9.6439   9.8857  10.3049  7.6865   \n",
       "\n",
       "              ZXDB     ZXDC   ZYG11B      ZYX    ZZEF1     ZZZ3  \n",
       "Patient_ID                                                       \n",
       "01CO001     8.4263   8.6865  10.0307   9.1699  10.2784  10.4263  \n",
       "01CO005     9.2946   9.1799   9.8408  10.5622  10.4512  10.4305  \n",
       "01CO006     8.7415   9.0741  10.3309  10.0512  10.2703  10.5430  \n",
       "01CO008     9.9542  10.0168  10.3365   9.9307  10.2538  10.8925  \n",
       "01CO013     8.7381   9.6165  10.3848  11.3066  11.2174  10.7616  \n",
       "...            ...      ...      ...      ...      ...      ...  \n",
       "21CO007     8.3707   8.8202   9.8234   9.4136   9.7245  10.6777  \n",
       "22CO004     9.0768   8.8857  10.2715   9.4737   9.6511  10.9114  \n",
       "22CO006     8.1447   8.7582  10.1497  10.4062  10.2491  10.3902  \n",
       "24CO005     7.8392   8.1849   9.6917   8.3264   7.8704  10.3740  \n",
       "27CO004     8.6110   9.1624   9.3106  11.0532   9.6777   9.6900  \n",
       "\n",
       "[106 rows x 13482 columns]"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# RENAME the path to analysis_data!\n",
    "protein_data.to_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_protein.csv\")\n",
    "protein_data_read_example = pd.read_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_protein.csv\", index_col=0)\n",
    "#protein_data_read_example\n",
    "\n",
    "# Repeat with clinical and RNA\n",
    "clinical_data.to_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_clinical.csv\")\n",
    "clinical_data_read_example = pd.read_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_clinical.csv\", index_col=0)\n",
    "#clinical_data_read_example\n",
    "\n",
    "# Repeat with clinical and RNA\n",
    "rna_data.to_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_rna.csv\")\n",
    "rna_data_read_example = pd.read_csv(\"/Users/peterchong/Documents/qbio_data_analysis_peterc/analysis_data/cptac/cptac_rna.csv\", index_col=0)\n",
    "#rna_data_read_example"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
